<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-Seq: Data Readiness for Machine Learning Applications: Data Collection: ArrayExpress</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/episode2.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-Seq: Data Readiness for Machine Learning Applications
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-Seq: Data Readiness for Machine Learning Applications
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-Seq: Data Readiness for Machine Learning Applications
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 9%" class="percentage">
    9%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 9%" aria-valuenow="9" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/episode2.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="episode1.html">1. Introduction: Machine Learning Ready RNA-Seq Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Data Collection: ArrayExpress
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#arrayexpress-file-formats">ArrayExpress File Formats</a></li>
<li><a href="#searching-for-a-dataset-on-arrayexpress">Searching for a Dataset on ArrayExpress</a></li>
<li><a href="#illustrative-dataset-ibd-dataset">Illustrative Dataset: IBD Dataset</a></li>
<li><a href="#downloading-and-reading-into-r">Downloading and Reading into R</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="episode3.html">3. Data Collection: GEO</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="episode4.html">4. Data Readiness: Data Format and Integrity</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="episode5.html">5. Data Readiness: Technical Noise</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="episode6.html">6. Data Readiness: Distribution and Scale</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode1.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode3.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode1.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction:
        </a>
        <a class="chapter-link float-end" href="episode3.html" rel="next">
          Next: Data Collection: GEO... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Data Collection: ArrayExpress</h1>
        <p>Last updated on 2024-03-05 |
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode2.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What format are processed RNA-Seq dataset stored on Array
Express?</li>
<li>How do I search for a dataset that meets my requirements on Array
Express?</li>
<li>Which files do I need to download?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain the expected data and file formats for processed RNA-Seq
datasets on ArrayExpress</li>
<li>Demonstrate ability to correctly identify the subset of files
required to create a dataset for a simple supervised machine learning
task (e.g., binary classification)</li>
<li>Demonstrate ability to download and save processed RNA-Seq data
files from ArrayExpress suitable for a simple supervised machine
learning task (e.g., binary classification)</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="arrayexpress-file-formats"><h2 class="section-heading">ArrayExpress File Formats<a class="anchor" aria-label="anchor" href="#arrayexpress-file-formats"></a>
</h2>
<hr class="half-width"><p>Array Express data sets comprise of four important file types that
contain all of the information about an experiment:</p>
<p><br></p>
<table class="table"><colgroup><col width="25%"><col width="75%"></colgroup><thead><tr class="header"><th>File Type</th>
<th>What It Is</th>
</tr></thead><tbody><tr class="odd"><td><strong>Sample and Data Relationship Format (SDRF)</strong></td>
<td>Tab-delimited file containing information about the samples, such as
phenotypical information, and factors that may be of interest in a
classification task. This file may also contain information about the
relationships between the samples, and details of the file names of the
raw data fastq for each sample.</td>
</tr><tr class="even"><td><strong>Investigation Description Format (IDF)</strong></td>
<td>Tab-delimited file containing top level information about the
experiment including title, description, submitter contact details and
protocols.</td>
</tr><tr class="odd"><td><strong>Raw data files</strong></td>
<td>Sequence data and quality scores, typically stored in FASTQ format.
There is however inconsistency - in some data sets, the term raw data
refers to processed data in the form of raw counts data, meaning RNA-Seq
counts that have not undergone further processing.</td>
</tr><tr class="even"><td><strong>Processed data files</strong></td>
<td>Processed data which includes raw counts or further processed
abundance measures such as TPM and FPKM. The data may be provided as a
separate file for each sample, or as a matrix of all samples</td>
</tr></tbody></table><p>Detailed information about BioStudies file formats is available in
the <a href="https://www.ebi.ac.uk/biostudies/misc/MAGE-TABv1.1_2011_07_28.pdf" class="external-link">MAGE-TAB
Specification document</a>.</p>
<p><br></p>
</section><section id="searching-for-a-dataset-on-arrayexpress"><h2 class="section-heading">Searching for a Dataset on ArrayExpress<a class="anchor" aria-label="anchor" href="#searching-for-a-dataset-on-arrayexpress"></a>
</h2>
<hr class="half-width"><p>We will use ArrayExpress to select an RNA-Seq dataset with a case
control design, suitable for constructing a machine learning
classification model. Machine learning based analyses generally perform
better the larger the sample size, and may perform very poorly and give
misleading results with insufficient samples. Given that, we’ll look for
a datasets with a relatively large number of samples.</p>
<p>Let’s begin on the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies" class="external-link">ArrayExpress
home page</a>.</p>
<p><br></p>
<figure><img src="fig/biostudies_home_page.jpg" alt="Screen Shot of Array Express Collection Tab" class="figure mx-auto d-block"><div class="figcaption">ArrayExpress Home Page</div>
</figure><p><br></p>
<p>We’ll apply the following filters to refine the results set, leaving
all other filters as the default with no selection:</p>
<p><br></p>
<table class="table"><thead><tr class="header"><th>Filter</th>
<th>Selection</th>
</tr></thead><tbody><tr class="odd"><td>Study type</td>
<td>rna-seq of coding rna</td>
</tr><tr class="even"><td>Organism</td>
<td>homo sapiens</td>
</tr><tr class="odd"><td>Technology</td>
<td>sequencing assay</td>
</tr><tr class="even"><td>Assay by Molecule</td>
<td>rna assay</td>
</tr><tr class="odd"><td>Processed Data Available</td>
<td>yes (check box)</td>
</tr></tbody></table><p><br></p>
<p>Now use the sort in the top right of the screen to sort the results
by “Samples”, in descending order. The results should look similar to
this:</p>
<p><br></p>
<figure><img src="fig/biostudies_home_filtered.jpg" alt="Screen Shot of Array Express Collection Tab Filtered Results" class="figure mx-auto d-block"><div class="figcaption">ArrayExpress Filtered Results</div>
</figure></section><section id="illustrative-dataset-ibd-dataset"><h2 class="section-heading">Illustrative Dataset: IBD Dataset<a class="anchor" aria-label="anchor" href="#illustrative-dataset-ibd-dataset"></a>
</h2>
<hr class="half-width"><p>In the search results displayed above, the top-ranked dataset is
E-MTAB-11349: <em>Whole blood expression profiling of patients with
inflammatory bowel diseases in the IBD-Character cohort</em>. We’ll call
it the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-11349" class="external-link">IBD
dataset</a>.</p>
<p>The IBD dataset comprises human samples of patients with inflammatory
bowel diseases and controls. The ‘Protocols’ section explains the main
steps in the generation of the data, from RNA extraction and sample
preparation, to sequencing and processing of the raw RNA-Seq data. The
nucleic acid sequencing protocol gives details of the sequencing
platform used to generate the raw fastq data, and the version of the
human genome used in the alignment step to generate the count data. The
normalization data transformation protocol gives the tools used to
normalise raw counts data for sequence depth and sample composition, in
this case normalisation is conducted using the R package <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" class="external-link">DESeq2</a>.
DESEq2 contains a range of functions for the transformation and analysis
of RNA-Seq data. Let’s look at some of the basic information on this
dataset:</p>
<p><br></p>
<table class="table"><thead><tr class="header"><th>Data Field</th>
<th>Values</th>
</tr></thead><tbody><tr class="odd"><td>Sample count</td>
<td>590</td>
</tr><tr class="even"><td>Experimental Design</td>
<td>case control design</td>
</tr><tr class="odd"><td>Experimental factors</td>
<td>Normal, Ulcerative colitis, Crohn’s disease</td>
</tr></tbody></table><p><br></p>
<p>The dataset contains two alternative sets of processed data, a matrix
of raw counts, and a matrix of processed counts that have been
normalised using DESeq2, and filtered to only include transcripts with a
mean raw expression read count &gt; 10.</p>
</section><section id="downloading-and-reading-into-r"><h2 class="section-heading">Downloading and Reading into R<a class="anchor" aria-label="anchor" href="#downloading-and-reading-into-r"></a>
</h2>
<hr class="half-width"><p>Let’s download the SDRF file and the raw counts matrix - these are
the two files that contain the information we will need to build a
machine-learning classification model. In the data files box to the
right-hand side, check the following two files
<code>E-MTAB-11349.sdrf.txt</code> and
<code>ArrayExpress-raw.csv</code>, and save them in a folder called
<code>data</code> in the working directory for your R project. For
consistency, rename <code>ArrayExpress-raw.csv</code> as
<code>E-MTAB-11349.counts.matrix.csv</code>.</p>
<p>For convenience, a copy of the files is also stored on zenodo. You
can run the following code that uses the function
<code>download.file()</code> to download the files and save them
directly into your <code>data</code> directory. (You need to have
created the <code>data</code> directory beforehand).</p>
<p>To get or set a working directory in R <code>getwd</code> function
can be used to return an absolute file path representing the current
working directory of the R process; if not correct <code>setwd</code>
can be used to set the working directory to the desired location.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/E-MTAB-11349.sdrf.txt"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/E-MTAB-11349.sdrf.txt"</span><span class="op">)</span></span>
<span>              </span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/E-MTAB-11349.counts.matrix.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/E-MTAB-11349.counts.matrix.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<div class="section level3">
<h3 id="raw-counts-matrix">Raw counts matrix<a class="anchor" aria-label="anchor" href="#raw-counts-matrix"></a></h3>
<p>In R studio, open your project workbook and read the raw counts data.
Then check the dimensions of the matrix to confirm we have the expected
number of samples and transcript IDs.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/E-MTAB-11349.counts.matrix.csv"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">","</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">writeLines</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i %s"</span>, <span class="fu">c</span><span class="op">(</span><span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"rows corresponding to transcript IDs"</span>, <span class="st">"columns corresponding to samples"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>22751 rows corresponding to transcript IDs
592 columns corresponding to samples</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>View a small subset of the data, (e.g. first ten rows and 8 columns)
to see how the samples are labelled and how the sample and transcript
IDs are formatted.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            read Sample 1 Sample 2 Sample 3 Sample 4 Sample 5 Sample 6
1   1          *    13961    16595    20722    17696    25703    20848
2   2 ERCC-00002        0        0        0        0        0        0
3   3 ERCC-00003        0        0        0        0        0        0
4   4 ERCC-00004        0        0        0        0        3        0
5   5 ERCC-00009        0        0        0        0        0        0
6   6 ERCC-00012        0        0        0        0        0        0
7   7 ERCC-00013        0        0        0        0        0        0
8   8 ERCC-00014        0        0        0        0        0        0
9   9 ERCC-00016        0        0        0        0        0        0
10 10 ERCC-00017        2        0        0        0        1        0</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sdrf-file">SDRF File<a class="anchor" aria-label="anchor" href="#sdrf-file"></a></h3>
<p>Now let’s read the sdrf file (a plain text file) into R and check the
dimensions of the file.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in the sdrf file</span></span>
<span></span>
<span><span class="va">samp.info.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/E-MTAB-11349.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i rows, corresponding to the samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 590 rows, corresponding to the samples"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i columns, corresponding to the available variables for each sample"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 32 columns, corresponding to the available variables for each sample"</code></pre>
</div>
<p>If we view the column names, we can see that the file does indeed
contain a set of variables describing both phenotypical and experimental
protocol information relating to each sample.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Source Name"                            
 [2] "Characteristics[organism]"              
 [3] "Characteristics[age]"                   
 [4] "Unit[time unit]"                        
 [5] "Term Source REF"                        
 [6] "Term Accession Number"                  
 [7] "Characteristics[developmental stage]"   
 [8] "Characteristics[sex]"                   
 [9] "Characteristics[individual]"            
[10] "Characteristics[disease]"               
[11] "Characteristics[organism part]"         
[12] "Material Type"                          
[13] "Protocol REF"                           
[14] "Protocol REF"                           
[15] "Protocol REF"                           
[16] "Extract Name"                           
[17] "Material Type"                          
[18] "Comment[LIBRARY_LAYOUT]"                
[19] "Comment[LIBRARY_SELECTION]"             
[20] "Comment[LIBRARY_SOURCE]"                
[21] "Comment[LIBRARY_STRATEGY]"              
[22] "Protocol REF"                           
[23] "Performer"                              
[24] "Assay Name"                             
[25] "Technology Type"                        
[26] "Protocol REF"                           
[27] "Derived Array Data File"                
[28] "Comment [Derived ArrayExpress FTP file]"
[29] "Protocol REF"                           
[30] "Derived Array Data File"                
[31] "Comment [Derived ArrayExpress FTP file]"
[32] "Factor Value[disease]"                  </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>ArrayExpress stores two standard files with information about each
experiment: 1. Sample and Data Relationship Format (SDRF) and 2.
Investigation Description Format (IDF).</li>
<li>ArrayExpress provides raw and processed data for RNA-Seq datasets,
typically stored as csv, tsv, or txt files.</li>
<li>The filters on the ArrayExpress website allow you to select and
download a dataset that suit your task.</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode1.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode3.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode1.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction:
        </a>
        <a class="chapter-link float-end" href="episode3.html" rel="next">
          Next: Data Collection: GEO... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode2.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:parkinsone@cardiff.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.2" class="external-link">sandpaper (0.16.2)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.3" class="external-link">pegboard (0.7.3)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.1" class="external-link">varnish (1.0.1)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode2.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "machine learning, RNA-Seq, data readiness, pre-processing, data management",
  "name": "Data Collection: ArrayExpress",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode2.html",
  "identifier": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode2.html",
  "dateCreated": "2023-05-23",
  "dateModified": "2024-03-05",
  "datePublished": "2024-03-05"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

