<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-Seq: Data Readiness for Machine Learning Applications: Data Readiness: Technical Noise</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/episode5.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-Seq: Data Readiness for Machine Learning Applications
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-Seq: Data Readiness for Machine Learning Applications
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-Seq: Data Readiness for Machine Learning Applications
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 80%" class="percentage">
    80%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 80%" aria-valuenow="80" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/episode5.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="episode1.html">1. Introduction: Machine Learning Ready RNA-Seq Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="episode2.html">2. Data Collection: ArrayExpress</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="episode3.html">3. Data Collection: GEO</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="episode4.html">4. Data Readiness: Data Format and Integrity</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Data Readiness: Technical Noise
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#technical-artefacts-in-rna-seq-data">Technical Artefacts in RNA-Seq Data</a></li>
<li><a href="#low-read-counts">Low read counts</a></li>
<li><a href="#outlier-reading-counts">Outlier Reading Counts</a></li>
<li><a href="#outlier-read-count-filtering">Outlier Read Count Filtering</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="episode6.html">6. Data Readiness: Distribution and Scale</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode4.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode6.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode4.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Readiness: Data
        </a>
        <a class="chapter-link float-end" href="episode6.html" rel="next">
          Next: Data Readiness:... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Data Readiness: Technical Noise</h1>
        <p>Last updated on 2024-01-09 |
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode5.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do technical artefacts in RNA-Seq data impact machine learning
algorithms?</li>
<li>How can technical artefacts such as low count genes and outlier read
counts be effectively removed from RNA-Seq data prior to analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Describe two important sources of technical noise in RNA-Seq data -
genes with very low read counts and genes with influential outlier read
counts - and recall why these exist in raw datasets</li>
<li>Demonstrate how to remove these noise elements from RNA-Seq count
data to further improve machine learning readiness using standard R
libraries</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="technical-artefacts-in-rna-seq-data"><h2 class="section-heading">Technical Artefacts in RNA-Seq Data<a class="anchor" aria-label="anchor" href="#technical-artefacts-in-rna-seq-data"></a>
</h2>
<hr class="half-width"><p>Machine learning classification algorithms are highly sensitive to
any feature data characteristic that differs between experimental
groups, and will exploit these differences when training a model. Given
this, it is important to make sure that data input into a machine
learning model reflects true biological signal, and not technical
artefacts or noise that stems from the experimental process used to
generate the data. In simple terms, we need to remove things are aren’t
“real biological information”.</p>
<p>There are two important sources of noise inherent in RNA-Seq data
that may negatively impact machine learning modelling performance,
namely low read counts, and influential outlier read counts.</p>
<p><br></p>
</section><section id="low-read-counts"><h2 class="section-heading">Low read counts<a class="anchor" aria-label="anchor" href="#low-read-counts"></a>
</h2>
<hr class="half-width"><p>Genes with consistently low read count values across all samples in a
dataset may be technical or biological stochastic artefacts such as the
detection of a transcript from a gene that is not uniformly active in a
heterogeneous cell population or as the result of a transcriptional
error. Below some threshold, differences in counts between the
conditions of interest for a given gene are unlikely to be
representative of true biological differences between the groups.
Filtering out low count genes has been show to increase the
classification performance of machine learning classifiers, and to
increase the stability of the set of genes selected by a machine
learning algorithm in the context of selecting relevant genes.</p>
<div class="section level3">
<h3 id="investigating-low-counts">Investigating Low Counts<a class="anchor" aria-label="anchor" href="#investigating-low-counts"></a></h3>
<p>We’ll briefly investigate the distribution of read counts in the IBD
dataset to illustrate this point. Import the cleaned up counts matrix
and sample information text files that we prepared in Episode 4. If you
didn’t manage to save them, you can download them directly using the
following code.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/ibd.sample.info.txt"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/ibd.sample.info.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/counts.mat.ibd.txt"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/counts.mat.ibd.txt"</span><span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<p>And now read the files into R…</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># suppressPackageStartupMessages(library(dplyr, quietly = TRUE))</span></span>
<span><span class="co"># suppressPackageStartupMessages(library(ggplot2, quietly = TRUE))</span></span>
<span><span class="co"># suppressPackageStartupMessages(library(tibble, quietly = TRUE))</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/ibd.sample.info.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts.mat.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/counts.mat.ibd.txt"</span>, sep<span class="op">=</span><span class="st">'\t'</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
<p>Run the following code to view the histogram giving the frequency of
the maximum count for each gene in the sample (plotted on a log10
scale). You’ll see in the resulting plot that over 800 genes have no
counts for any gene (i.e., maximum = 0), and that there are hundreds of
genes where the maximum count for the gene across all samples is below
10. For comparison, the median of these maximum counts across all genes
is around 250. These low count genes are likely to represent technical
noise.</p>
<p>A simple filtering approach is to remove all genes where the maximum
read count for that gene over all samples is below a given threshold.
The next step is to determine what this threshold should be.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">`%&gt;%`</span> <span class="op">&lt;-</span> <span class="fu">magrittr</span><span class="fu">::</span><span class="va">`%&gt;%`</span></span>
<span></span>
<span><span class="fu">data.frame</span><span class="op">(</span>max_count <span class="op">=</span> <span class="fu">apply</span><span class="op">(</span><span class="va">counts.mat.ibd</span>, <span class="fl">1</span>, <span class="va">max</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">max_count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">200</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">xlab</span><span class="op">(</span><span class="st">"Max Counts (log10 scale)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ylab</span><span class="op">(</span><span class="st">"Frequency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">scale_x_log10</span><span class="op">(</span>n.breaks <span class="op">=</span> <span class="fl">6</span>, labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va">comma</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Removed 10 rows containing non-finite values (`stat_bin()`).</code></pre>
</div>
<figure><img src="fig/episode5-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>Based on the distribution of the maximum read counts per gene
displayed in the histogram above, where would you estimate the cutoff
for low count genes should be set?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>There is no right answer. However, looking at the histogram, you can
see an approximately symmetrical distribution centred around a value
between 100 and 1,000 (approximately 250), and a second distirbution of
low count genes in the range 1-20. From this, we could estimate that the
cut off for low count genes will be approximately in the range
10-20.</p>
</div>
</div>
</div>
</div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="read-count-normalisation">Read Count Normalisation<a class="anchor" aria-label="anchor" href="#read-count-normalisation"></a></h3>
<p>In order to be able to compare read counts between samples, we must
first adjust (‘normalise’) the counts to control for differences in
sequence depth and sample composition between samples. To achieve this,
we run the following code that will normalise the counts matrix using
the median-of-ratios method implemented in the R package
<code>DESeq2</code>. For more information on the rationale for scaling
RNA-Seq counts and a comparison of the different metrics used see <a href="https://www.youtube.com/watch?v=tO2H3zuBouw" class="external-link">RDMBites | RNAseq
expression data</a>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert the condition variable to a factor as required by DESeq2</span></span>
<span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'condition'</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'condition'</span><span class="op">)</span><span class="op">]</span>, <span class="va">factor</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create DESeq Data Set object from the raw counts (converted to a matrix), with condition as the factor of interest</span></span>
<span><span class="va">dds.ibd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>    countData <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>, row.names <span class="op">=</span> <span class="st">'sampleID'</span><span class="op">)</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate the normalised count values using the median-of-ratios method</span></span>
<span><span class="va">dds.ibd</span> <span class="op">&lt;-</span> <span class="va">dds.ibd</span> <span class="op">%&gt;%</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the normalised counts</span></span>
<span><span class="va">counts.ibd.norm</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">counts</span><span class="op">(</span><span class="va">dds.ibd</span>, normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="setting-a-low-counts-threshold">Setting a Low Counts Threshold<a class="anchor" aria-label="anchor" href="#setting-a-low-counts-threshold"></a></h3>
<p>There are a number of methodologies to calculate the appropriate
threshold. One widely used approach calculates the threshold that
maximises the similarity between the samples, calculating the mean
pairwise Jaccard index over all samples in each conditionn of interest,
and setting the threshold that maximises this value. The motivation for
this measure is that true biological signal will be consistent between
samples in the same condition, however low count noise will be randomly
distributed. By maximising the similarity between samples in the same
class, the threshold is likely to be set above the background noise
level. The code for this filter is implemented in the R package
<code>HTSFilter</code> and the methodology, along with details of
alternative approaches, is described in detail by <a href="10.1093/bioinformatics/btt350">Rau et al., 2013</a>.</p>
<p>Calculating the Jaccard index between all pairs of samples in a
dataset does not however scale well with the dimensionality of the data.
Here we use an alternative approach that achieves a very similar result
using the Multiset Jaccard index, which is faster to compute. Don’t
worry too much about the code, the main point is that we find a filter
threshold, and filter out all the genes where the max count value is
below the threshold, on the basis that these are likely technical noise.
Run the code to plot the Multiset Jaccard Index for a series of values
for the filter threshold.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For the purpose of illustration, and to shorten the run time, we sample 5,000 genes from the counts matrix.</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts.ibd.norm.samp</span> <span class="op">&lt;-</span> <span class="va">counts.ibd.norm</span><span class="op">[</span><span class="fu">sample</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">counts.ibd.norm</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5000</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># create vector of the class labels</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span></span>
<span></span>
<span><span class="co"># create a sequence of thresholds to test</span></span>
<span><span class="va">t.seq</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">25</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Function to calculate the Multiset Jaccard Index' over all samples</span></span>
<span><span class="va">Jaccard</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">row.sums</span> <span class="op">=</span> <span class="fu">rowSums</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span>  <span class="va">inter</span> <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">row.sums</span> <span class="op">==</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">union</span> <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">row.sums</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">inter</span><span class="op">/</span><span class="va">union</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Calculate vector of the minimum Multiset Jaccard Index over the three classes, for each value of t</span></span>
<span></span>
<span><span class="va">ms.jac</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">t.seq</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">filt.counts</span> <span class="op">=</span> <span class="va">counts.ibd.norm.samp</span> <span class="op">&gt;=</span><span class="va">t</span></span>
<span>  <span class="va">group.jac</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">condition</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cond</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu">Jaccard</span><span class="op">(</span><span class="va">filt.counts</span><span class="op">[</span>,<span class="va">condition</span><span class="op">==</span><span class="va">cond</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">group.jac</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># plot the threshold value against the value of the Multiset Jaccard index to visualise</span></span>
<span></span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ggplot</span><span class="op">(</span>data<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>t <span class="op">=</span> <span class="va">t.seq</span>, jacc <span class="op">=</span> <span class="va">ms.jac</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">geom_line</span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">t</span>, y<span class="op">=</span><span class="va">jacc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu">max</span><span class="op">(</span><span class="va">ms.jac</span><span class="op">)</span>, lty<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="st">'gray'</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">geom_point</span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="fu">which.max</span><span class="op">(</span><span class="va">ms.jac</span><span class="op">)</span>, y<span class="op">=</span><span class="fu">max</span><span class="op">(</span><span class="va">ms.jac</span><span class="op">)</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"red"</span>, cex<span class="op">=</span><span class="fl">6</span>, pch<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">xlab</span><span class="op">(</span><span class="st">"Low Count Threshold"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ylab</span><span class="op">(</span><span class="st">"Multiset Jaccard Index"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/episode5-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2:<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<p>What value should we use for the low counts threshold?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The threshold value is given by the following code, which should
return a value close to 10 for this dataset.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">t.hold</span> <span class="op">&lt;-</span> <span class="fu">which.max</span><span class="op">(</span><span class="va">ms.jac</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 11</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="filtering-low-counts">Filtering Low Counts<a class="anchor" aria-label="anchor" href="#filtering-low-counts"></a></h3>
<p>Having determined a threshold, we then filter the raw counts matrix
on the rows (genes) that meet the threshold criterion based on the
normalised counts. As you can see, around 4,000 genes are removed from
the dataset, which is approximately 20% of the genes. These genes are
unlikely to contain biologically meaningful information but they might
easily bias a machine learning classifier.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.ibd.filtered</span> <span class="op">&lt;-</span> <span class="va">counts.mat.ibd</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">counts.ibd.norm</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu">sum</span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="va">t.hold</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Genes filtered: %s; Genes remaining: %s"</span>, <span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">-</span><span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd.filtered</span><span class="op">)</span>, <span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd.filtered</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Genes filtered: 3712; Genes remaining: 19038"</code></pre>
</div>
</div>
</section><section id="outlier-reading-counts"><h2 class="section-heading">Outlier Reading Counts<a class="anchor" aria-label="anchor" href="#outlier-reading-counts"></a>
</h2>
<hr class="half-width"><p>A second source of technical noise is outlier read counts. Relatively
high read counts occurring in only a very small number of samples
relative to the size of each patient group are unlikely to be
representative of the general population and a result of biological
heterogeneity and technical effects. These very large count values may
bias machine learning algorithms if they help to discriminate between
examples in a training set, leading to model overfitting. Such
influential outliers may be the result of natural variation between
individuals or they may have been introduced during sample preparation.
cDNA libraries require PCR amplification prior to sequencing to achieve
sufficient sequence depth. This clonal amplification by PCR is
stochastic in nature; different fragments may be amplified with
different probabilities. This leaves the possibility of outlier read
counts having resulted from bias introduced by amplification, rather
than biological differences between samples. The presence of these
outlier read counts for a particular gene may therefore inflate the
observed association between a particular gene and the condition of
interest.</p>
<p>Run the following code to view the top 10 values of read counts in
the raw counts matrix. Compare this with the histogram above, and the
mean read count. The largest read count values range from 2MM to over
3MM counts for a gene in a particular sample. These require
investigation!</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">)</span>,<span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 2037946 2038514 2043983 2133125 2238093 2269033 2341479 2683585 3188911
[10] 3191428</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"The mean read count value: %f"</span>, <span class="fu">mean</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "The mean read count value: 506.731355"</code></pre>
</div>
</section><section id="outlier-read-count-filtering"><h2 class="section-heading">Outlier Read Count Filtering<a class="anchor" aria-label="anchor" href="#outlier-read-count-filtering"></a>
</h2>
<hr class="half-width"><p>As with low counts, multiple methods exist to identify influential
outlier read counts. A summary is provided by <a href="https://doi.org/10.3389/fgene.2023.1158352" class="external-link">Parkinson et al.</a>.
Here we adopt an approach that is built into the R package
<code>DESeq2</code> that identifies potential influential outiers based
on their Cooks distance.</p>
<p>Run the following code to create DESeq Data Set object from the
filtered raw counts matrix, with condition as the experimental factor of
interest.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds.ibd.filt</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span></span>
<span>    countData <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts.mat.ibd.filtered</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>, row.names <span class="op">=</span> <span class="st">'sampleID'</span><span class="op">)</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">condition</span><span class="op">)</span></span></code></pre>
</div>
<p>Run <code>DESeq2</code> differential expression analysis, which
automatically calculates the Cook’s distances for every read count. This
may take a few minutes to run.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deseq.ibd</span> <span class="op">&lt;-</span>  <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeq</span><span class="op">(</span><span class="va">dds.ibd.filt</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating size factors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-- replacing outliers and refitting for 1559 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cooks.mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu">assays</span><span class="op">(</span><span class="va">deseq.ibd</span><span class="op">)</span><span class="op">[[</span><span class="st">"cooks"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<p>We now calculate the cooks outlier threshold by computing the
expected F-distribution for the number of samples in the dataset.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cooks.quantile</span> <span class="op">&lt;-</span> <span class="fl">0.95</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">deseq.ibd</span><span class="op">)</span>     <span class="co"># number of samples</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">3</span>                   <span class="co"># number of model parameters (in the three condition case)</span></span>
<span></span>
<span><span class="va">h.threshold</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu">qf</span><span class="op">(</span><span class="va">cooks.quantile</span>, <span class="va">p</span>, <span class="va">m</span> <span class="op">-</span> <span class="va">p</span><span class="op">)</span></span></code></pre>
</div>
<p>Filter the counts matrix to eliminate all genes where the cooks
distance is over the outlier threshold. Here you can see that a further
~1,800 genes are filtered out based on having outlier read count values
for at least one sample.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.ibd.ol.filtered</span> <span class="op">&lt;-</span>  <span class="va">counts.mat.ibd.filtered</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">cooks.mat</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="op">(</span><span class="fu">max</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">h.threshold</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Genes filtered: %s; Genes remaining: %s"</span>, <span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd.filtered</span><span class="op">)</span><span class="op">-</span><span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd.ol.filtered</span><span class="op">)</span>, <span class="fu">nrow</span><span class="op">(</span><span class="va">counts.mat.ibd.ol.filtered</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Genes filtered: 1776; Genes remaining: 17262"</code></pre>
</div>
<p>Run the same plot as above, and we can see that the genes with very
low maximum counts over all samples have been removed. Note that here we
are looking at the raw unnormalised counts, so not all maximum counts
are below 11, as the filter was applied to the normailsed counts.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data.frame</span><span class="op">(</span>max_count <span class="op">=</span> <span class="fu">apply</span><span class="op">(</span><span class="va">counts.mat.ibd.ol.filtered</span>, <span class="fl">1</span>, <span class="va">max</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">max_count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">200</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">xlab</span><span class="op">(</span><span class="st">"Max Counts (log10 scale)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">ylab</span><span class="op">(</span><span class="st">"Frequency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu">scale_x_log10</span><span class="op">(</span>n.breaks <span class="op">=</span> <span class="fl">6</span>, labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va">comma</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/episode5-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><br></p>
<p>Now save the output as a file that we will use in the next
episode.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">counts.mat.ibd.ol.filtered</span>, file<span class="op">=</span><span class="st">"./data/counts.mat.ibd.ol.filtered.txt"</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>RNA-Seq read counts contain two main sources of technical ‘noise’
that are unlikely to represent true biological information, and may be
artefacts from experimental processes required to generate the data: Low
read counts and influential outlier read counts.</li>
<li>Filtering out low count and influential outlier genes removes
potentially biasing variables without negatively impacting the
performance of downstream machine learning analysis. Gene filtering is
therefore an important step in preparing an RNA-Seq dataset to be
machine learning ready.</li>
<li>Multiple approaches exist to identify the specific genes with
uninformative low count and influential outlier read counts in RNA-Seq
data, however they all aim to find the boundary between true biological
signal and technical and biological artefacts.</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode4.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode6.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode4.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Readiness: Data
        </a>
        <a class="chapter-link float-end" href="episode6.html" rel="next">
          Next: Data Readiness:... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode5.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:parkinsone@cardiff.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.2" class="external-link">sandpaper (0.16.2)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.3" class="external-link">pegboard (0.7.3)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.1" class="external-link">varnish (1.0.1)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode5.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "machine learning, RNA-Seq, data readiness, pre-processing, data management",
  "name": "Data Readiness: Technical Noise",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode5.html",
  "identifier": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode5.html",
  "dateCreated": "2023-05-23",
  "dateModified": "2024-01-09",
  "datePublished": "2024-01-09"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

