<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-Seq: Data Readiness for Machine Learning Applications: Data Readiness: Data Format and Integrity</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/episode4.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-Seq: Data Readiness for Machine Learning Applications
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-Seq: Data Readiness for Machine Learning Applications
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-Seq: Data Readiness for Machine Learning Applications
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 38%" class="percentage">
    38%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 38%" aria-valuenow="38" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/episode4.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="episode1.html">1. Introduction: Machine Learning Ready RNA-Seq Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="episode2.html">2. Data Collection: ArrayExpress</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="episode3.html">3. Data Collection: GEO</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Data Readiness: Data Format and Integrity
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#required-format-for-machine-learning-libraries">Required format for machine learning libraries</a></li>
<li><a href="#idb-dataset-sample-information-clean-up">IDB dataset: Sample Information Clean Up</a></li>
<li><a href="#idb-dataset-counts-matrix-clean-up">IDB dataset: Counts Matrix Clean Up</a></li>
<li><a href="#the-tb-dataset-another-example">The TB Dataset: Another Example</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="episode5.html">5. Data Readiness: Technical Noise</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="episode6.html">6. Data Readiness: Distribution and Scale</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode3.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode5.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode3.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Collection: GEO
        </a>
        <a class="chapter-link float-end" href="episode5.html" rel="next">
          Next: Data Readiness:... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Data Readiness: Data Format and Integrity</h1>
        <p> Last updated on 2023-11-10 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode4.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What format do I require my data in to build a supervised machine
learning classifier?</li>
<li>What are the potential data format and data integrity issues I will
encounter with RNA-Seq datasets, including those downloaded from public
repositories, that I’ll need to address before beginning any
analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Describe the format of a dataset required as input to a simple
supervised machine learning analysis</li>
<li>Recall a checklist of common formatting and data integrity issues
encountered with sample information (metadata) in RNA-Seq datasets</li>
<li>Recall a checklist of common formatting and data integrity issues
encountered with counts matrix RNA-Seq datasets</li>
<li>Recall why formatting and data integrity issues present a problem
for downstream ML analysis</li>
<li>Apply the checklist to an unseen dataset downloaded from either the
Array Express or GEO platform to identify potential issues</li>
<li>Construct a reusable R code pipeline implement required changes to a
new dataset</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="required-format-for-machine-learning-libraries"><h2 class="section-heading">Required format for machine learning libraries<a class="anchor" aria-label="anchor" href="#required-format-for-machine-learning-libraries"></a>
</h2>
<hr class="half-width"><p>Data must be correctly formatted for use in a machine learning / AI
pipeline. The <em>garbage in garbage out</em> principle applies; a
machine learning model is only as good as the input data. This episode
first discusses the kind of data required as input to a machine learing
classification model. We’ll then illustrates the process of formatting
an RNA-Seq dataset downloaded from a public repository. On the way,
we’ll define a checklist of data integrity issues to watch out for.</p>
<p>For a supervised machine learning classification task, we
require:</p>
<ol style="list-style-type: decimal"><li><p><strong>A matrix of values of all of the predictor
variables</strong> to be included in a machine learning model for each
of the samples. Example predictor variables would be gene abundance in
the form of read counts. Additional predictor variables may be obtained
from the sample information including demographic data
(e.g. <code>sex</code>, <code>age</code>) and clinical and laboratory
measures.</p></li>
<li><p><strong>A vector of values for the target variable</strong> that
we are trying to predict for each sample. Example target variables for a
classification model would be disease state, for example ‘infected with
tuburculosis’ compared with ‘healthy control’. The analogy for a
regression task would be a continuous variable such as a disease
progression or severity score.</p></li>
</ol><p>We will follow a series of steps to reformat and clean up the sample
information and counts matrix file for the IBD dataset collected form
ArrayExpress in Episode 2 by applying a checklist of data clean up
items. If required, the data can be downloaded again by running this
code.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/E-MTAB-11349.sdrf.txt"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/E-MTAB-11349.sdrf.txt"</span><span class="op">)</span></span>
<span>              </span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/E-MTAB-11349.counts.matrix.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/E-MTAB-11349.counts.matrix.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>We’ll start by reading in the dataset from our <code>data</code>
subfolder.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-11349.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">raw.counts.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-11349.counts.matrix.csv"</span>, sep<span class="op">=</span><span class="st">","</span> , header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="idb-dataset-sample-information-clean-up"><h2 class="section-heading">IDB dataset: Sample Information Clean Up<a class="anchor" aria-label="anchor" href="#idb-dataset-sample-information-clean-up"></a>
</h2>
<hr class="half-width"><p>In real RNA-Seq experiments, sample information (metadata) has likely
been collected from a range of sources, for example medical records,
clinical and laboratory readings and then manually compiled in a matrix
format, often a spreadsheet, before being submitted to a public
respository. Each of these steps has the potential to introduce errors
and inconsistency in the data that must be investigated prior to any
downstream machine learning analysis. Whether you are working with
sample information received directly from co-workers or downloaded from
a public repsository, the following guidelines will help identify and
resolve issues that may impact downstream analysis.</p>
<div id="sample-information-readiness-checklist" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="sample-information-readiness-checklist" class="callout-inner">
<h3 class="callout-title">Sample Information Readiness Checklist<a class="anchor" aria-label="anchor" href="#sample-information-readiness-checklist"></a>
</h3>
<div class="callout-content">
<table class="table"><colgroup><col width="7%"><col width="42%"><col width="50%"></colgroup><thead><tr class="header"><th>Item</th>
<th>Check For…</th>
<th>Rationale</th>
</tr></thead><tbody><tr class="odd"><td>1</td>
<td>
<strong>Unique identifier</strong> matching between counts matrix
and sample information</td>
<td>Target variables must be matched to predictor variables for each
sample. Target variables are the thing we are trying to predict such as
disease state. Predictor variables are the things we are going to
predict based on, such as the expression levels of particular
genes.</td>
</tr><tr class="even"><td>2</td>
<td>
<strong>Appropriate target variable</strong> present in the
data</td>
<td>The sample information needs to contain the target variable you are
trying to predict with the machine learning model</td>
</tr><tr class="odd"><td>3</td>
<td>
<strong>Unique variable names</strong>, without spaces or special
characters</td>
<td>Predictor variables may be drawn from the sample information. ML
algorithms require unique variables. Spaces etc. may cause errors with
bioinformatics and machine learning libraries used downstream</td>
</tr><tr class="even"><td>4</td>
<td>
<strong>Machine readable and consistent encoding</strong> of
categorical variables</td>
<td>Inconsistent spellings between samples will be treated as separate
values. Application specific formatting such as colour coding and notes
in spreadsheets will be lost when converting to plain text formats.</td>
</tr><tr class="odd"><td>5</td>
<td>
<strong>Numerical class values</strong> formatted as a factor (for
classification)</td>
<td>Some machine learning algorithms/libraries (e.g. SVM) require
classes defined by a number (e.g. -1 and +1 )</td>
</tr><tr class="even"><td>6</td>
<td>
<strong>Balance of classes</strong> in the target variable (for
classification)</td>
<td>Imbalanced datasets may negatively impact results of machine
learning algorithms</td>
</tr></tbody></table></div>
</div>
</div>
<p><br></p>
<p>We’ll now apply these steps sequentially to the sample information
for the IBD dataset, contained in our variable
<code>samp.info.ibd</code>. First let’s take a look at the data:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 32
$ `Source Name`                             &lt;chr&gt; "Sample 1", "Sample 2", "Sam…
$ `Characteristics[organism]`               &lt;chr&gt; "Homo sapiens", "Homo sapien…
$ `Characteristics[age]`                    &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 2…
$ `Unit[time unit]`                         &lt;chr&gt; "year", "year", "year", "yea…
$ `Term Source REF`                         &lt;chr&gt; "EFO", "EFO", "EFO", "EFO", …
$ `Term Accession Number`                   &lt;chr&gt; "UO_0000036", "UO_0000036", …
$ `Characteristics[developmental stage]`    &lt;chr&gt; "adult", "adult", "adult", "…
$ `Characteristics[sex]`                    &lt;chr&gt; "male", "female", "male", "m…
$ `Characteristics[individual]`             &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 1…
$ `Characteristics[disease]`                &lt;chr&gt; "Crohns disease\tblood\torga…
$ `Characteristics[organism part]`          &lt;chr&gt; "", "blood", "blood", "blood…
$ `Material Type`                           &lt;chr&gt; "", "organism part", "organi…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117946", "P-MTAB…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117947", "P-MTAB…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117948", "P-MTAB…
$ `Extract Name`                            &lt;chr&gt; "", "Sample 2", "Sample 3", …
$ `Material Type`                           &lt;chr&gt; "", "RNA", "RNA", "RNA", "RN…
$ `Comment[LIBRARY_LAYOUT]`                 &lt;chr&gt; "", "PAIRED", "PAIRED", "PAI…
$ `Comment[LIBRARY_SELECTION]`              &lt;chr&gt; "", "other", "other", "other…
$ `Comment[LIBRARY_SOURCE]`                 &lt;chr&gt; "", "TRANSCRIPTOMIC", "TRANS…
$ `Comment[LIBRARY_STRATEGY]`               &lt;chr&gt; "", "Targeted-Capture", "Tar…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117950", "P-MTAB…
$ Performer                                 &lt;chr&gt; "", "Wellcome Trust Clinical…
$ `Assay Name`                              &lt;chr&gt; "", "Sample 2", "Sample 3", …
$ `Technology Type`                         &lt;chr&gt; "", "sequencing assay", "seq…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117949", "P-MTAB…
$ `Derived Array Data File`                 &lt;chr&gt; "", "ArrayExpress-normalized…
$ `Comment [Derived ArrayExpress FTP file]` &lt;chr&gt; "", "ftp://ftp.ebi.ac.uk/pub…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117949", "P-MTAB…
$ `Derived Array Data File`                 &lt;chr&gt; "", "ArrayExpress-raw.csv", …
$ `Comment [Derived ArrayExpress FTP file]` &lt;chr&gt; "", "ftp://ftp.ebi.ac.uk/pub…
$ `Factor Value[disease]`                   &lt;chr&gt; "", "normal", "normal", "nor…</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>On first glance, which of our checklist items do you think could be
an issue in the sample information for this dataset?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>All of the checklist items apply in this case. Let’s go through
them…</p>
</div>
</div>
</div>
</div>
<p><br></p>
<ol style="list-style-type: decimal"><li>Let’s verify that there is a unique identifier that matches between
the sample information and counts matrix, and identify the name of the
column in the sample information. We manually find the names of the
samples in <code>raw.counts.ibd</code> (the counts matrix), and set to a
new variable <code>ibd.samp.names</code>. We then write a
<code>for loop</code> that evaluates which columns in the sample
information match the sample names. There is one match, the column named
<code>Source Name</code>. Note that there are other columns such as
<code>Assay Name</code> in this dataset that contain identifiers for
some but not all samples. This is a good illustration of why it is
important to check carefully to ensure you have a complete set of unique
identifiers.</li>
</ol><div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ibd.samp.names</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">lst.colnames</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_along</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">lst.colnames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">all</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span> <span class="op">==</span> <span class="va">ibd.samp.names</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"The unique IDs that match the counts matrix are in column: %s"</span>, <span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">lst.colnames</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "The unique IDs that match the counts matrix are in column: Source Name"</code></pre>
</div>
<p><br></p>
<ol start="2" style="list-style-type: decimal"><li>We next extract this unique identifier along with the condition of
interest, (in this case the <code>disease</code>), and any other
variables that may be potential predictor variables or used to assess
confounding factors in the subsequent analysis of the data (in this case
patient <code>sex</code> and <code>age</code>). We will use functions
from the <code>dplyr</code> package, part of the <code>tidyverse</code>
collection of packages throughout this episode.</li>
</ol><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># suppressPackageStartupMessages(library(dplyr, quietly = TRUE))</span></span>
<span><span class="co"># suppressPackageStartupMessages(library(ggplot2, quietly = TRUE))</span></span>
<span><span class="co"># suppressPackageStartupMessages(library(tibble, quietly = TRUE))</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">samp.info.ibd</span>,</span>
<span>                              <span class="st">'Source Name'</span>,</span>
<span>                              <span class="st">'Characteristics[age]'</span>,</span>
<span>                              <span class="st">'Characteristics[sex]'</span>,</span>
<span>                              <span class="st">'Characteristics[disease]'</span></span>
<span>                              <span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="3" style="list-style-type: decimal"><li>Let’s then rename the variables to something more easy to interpret
for humans, avoiding spaces and special characters. We’ll save these
selected columns to a new variable name
<code>samp.info.ibd.sel</code>.</li>
</ol><div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>,</span>
<span>                          <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Source Name'</span>,</span>
<span>                          <span class="st">'age'</span> <span class="op">=</span> <span class="st">'Characteristics[age]'</span>,</span>
<span>                          <span class="st">'sex'</span> <span class="op">=</span> <span class="st">'Characteristics[sex]'</span>,</span>
<span>                          <span class="st">'condition'</span> <span class="op">=</span> <span class="st">'Characteristics[disease]'</span></span>
<span>                          <span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="4" style="list-style-type: decimal"><li>Now check how each of the variables are encoded in our reduced
sample information data to identify any errors, data gaps and
inconsistent coding of categorical variables. Firstly let’s take a look
at the data.</li>
</ol><div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 4
$ sampleID  &lt;chr&gt; "Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5", …
$ age       &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 24, 39, 41, 38, 46, 7, 46, 21, 35…
$ sex       &lt;chr&gt; "male", "female", "male", "male", "female", "female", "femal…
$ condition &lt;chr&gt; "Crohns disease\tblood\torganism part\tP-MTAB-117946\tP-MTAB…</code></pre>
</div>
<p><br></p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2:<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<p>What steps do we need to take to ensure that these four data columns
are consistently formatted and machine readable? Are you able to write
the code to address these issues?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ul><li>There is clearly an issue with the coding of the
<code>condition</code> Crohn’s disease. We’ll fix that first, and then
check the consistency of the coding of categorical variables
<code>sex</code> and <code>condition</code>.</li>
</ul><div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">[</span><span class="fu">agrep</span><span class="op">(</span><span class="st">"Crohns"</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"crohns_disease"</span></span>
<span></span>
<span><span class="fu">unique</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">sex</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "male"               "female"             "crohns_disease"    
[4] "normal"             "ulcerative colitis"</code></pre>
</div>
<ul><li>Both <code>sex</code> and <code>condition</code> are consistently
encoded over all samples. We should however remove the spaces in the
values for <code>sampleID</code> and for the <code>condition</code>
value ulcerative colitis. Note: since we are changing our unique
identifier, we’ll need to make the same update later in the counts
matrix.</li>
</ul><p>Before we do this, since we’ll be using the pipe operator from the
tidyverse libraries, we will to define the pipe operator from the
magrittr package in the environment.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">`%&gt;%`</span> <span class="op">&lt;-</span> <span class="fu">magrittr</span><span class="fu">::</span><span class="va">`%&gt;%`</span></span>
<span></span>
<span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="va">samp.info.ibd.sel</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>sampleID <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">sampleID</span><span class="op">)</span>,</span>
<span>                                      condition <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="5" style="list-style-type: decimal"><li>As some machine learning classification algorithms such as support
vector machines require a numerical input for each class, we’ll add in a
new numerical column to represent our target variable called
<code>class</code> where all patients denoted as <code>normal</code> are
given the value -1, and all patients denoted as either
<code>crohns_disease</code> or <code>ulcerative_colitis</code> are given
the vale +1. We’ll also now convert the type of categorical columns to
factors as this is preferred by downstream libraries used for data
normalisation such as DESeq2.</li>
</ol><div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="va">samp.info.ibd.sel</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>class <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">if_else</span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">'normal'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'sex'</span>, <span class="st">'condition'</span>, <span class="st">'class'</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'sex'</span>, <span class="st">'condition'</span>, <span class="st">'class'</span><span class="op">)</span><span class="op">]</span>, <span class="va">factor</span><span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="6" style="list-style-type: decimal"><li>Finally, let’s check the distribution of the classes by creating a
<code>table</code> from the <code>class</code> column.</li>
</ol><div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">class</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
 -1   1 
267 323 </code></pre>
</div>
<p><br></p>
<p>The two classes are approximately equally represented, so let’s check
everything one last time.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 5
$ sampleID  &lt;chr&gt; "Sample_1", "Sample_2", "Sample_3", "Sample_4", "Sample_5", …
$ age       &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 24, 39, 41, 38, 46, 7, 46, 21, 35…
$ sex       &lt;fct&gt; male, female, male, male, female, female, female, male, male…
$ condition &lt;fct&gt; crohns_disease, normal, normal, normal, normal, ulcerative_c…
$ class     &lt;fct&gt; 1, -1, -1, -1, -1, 1, 1, -1, 1, 1, -1, 1, -1, -1, -1, 1, -1,…</code></pre>
</div>
<p><br></p>
<p>And save the output as a file.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>, file<span class="op">=</span><span class="st">"./data/ibd.sample.info.txt"</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="idb-dataset-counts-matrix-clean-up"><h2 class="section-heading">IDB dataset: Counts Matrix Clean Up<a class="anchor" aria-label="anchor" href="#idb-dataset-counts-matrix-clean-up"></a>
</h2>
<hr class="half-width"><div id="count-matrix-readiness-checklist" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="count-matrix-readiness-checklist" class="callout-inner">
<h3 class="callout-title">Count Matrix Readiness Checklist<a class="anchor" aria-label="anchor" href="#count-matrix-readiness-checklist"></a>
</h3>
<div class="callout-content">
<table class="table"><colgroup><col width="7%"><col width="42%"><col width="50%"></colgroup><thead><tr class="header"><th>Item</th>
<th>Check For…</th>
<th>Rationale</th>
</tr></thead><tbody><tr class="odd"><td>1</td>
<td>
<strong>Extraneous information</strong> that you don’t want to
include as a predictor variable</td>
<td>Blank rows, alternative gene/ transcript IDs and metadata about the
genes/transcripts will be treated as a predictor variable unless you
explicitly remove them</td>
</tr><tr class="even"><td>2</td>
<td>
<strong>Unique variable names</strong>, without spaces or special
characters</td>
<td>ML algorithms require unique variables. Spaces etc. may cause errors
with bioinformatics and ml libraries used downstream (and they must
match the sample information)</td>
</tr><tr class="odd"><td>3</td>
<td><strong>Duplicate data</strong></td>
<td>Duplicate variables may bias results of analysis and should be
removed unless intentional</td>
</tr><tr class="even"><td>4</td>
<td><strong>Missing values</strong></td>
<td>Missing values may cause errors and bias results. Missing values
must be identified and appropriate treatment determined (e.g. drop vs
impute)</td>
</tr></tbody></table></div>
</div>
</div>
<p><br></p>
<ol style="list-style-type: decimal"><li>Take a look at a sample of columns and rows to see what the
downloaded file looks like</li>
</ol><div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            read Sample 1 Sample 2 Sample 3 Sample 4 Sample 5 Sample 6
1   1          *    13961    16595    20722    17696    25703    20848
2   2 ERCC-00002        0        0        0        0        0        0
3   3 ERCC-00003        0        0        0        0        0        0
4   4 ERCC-00004        0        0        0        0        3        0
5   5 ERCC-00009        0        0        0        0        0        0
6   6 ERCC-00012        0        0        0        0        0        0
7   7 ERCC-00013        0        0        0        0        0        0
8   8 ERCC-00014        0        0        0        0        0        0
9   9 ERCC-00016        0        0        0        0        0        0
10 10 ERCC-00017        2        0        0        0        1        0</code></pre>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3:<a class="anchor" aria-label="anchor" href="#challenge-3"></a>
</h3>
<div class="callout-content">
<p>Can you see the irrelevant information that we need to remove from
the counts matrix right away?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<ul><li>Remove the first row, which is in this case totals of counts for
each sample</li>
<li>Remove the first columns, which just numbers the transcript IDs</li>
<li>Move the column named <code>read</code> that contains to the
transcript IDs to the row names</li>
</ul><div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.ibd</span> <span class="op">&lt;-</span> <span class="va">raw.counts.ibd</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="va">counts.mat.ibd</span> <span class="op">&lt;-</span>  <span class="va">counts.mat.ibd</span> <span class="op">%&gt;%</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'read'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts.mat.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           Sample 1 Sample 2 Sample 3 Sample 4 Sample 5 Sample 6
ERCC-00002        0        0        0        0        0        0
ERCC-00003        0        0        0        0        0        0
ERCC-00004        0        0        0        0        3        0
ERCC-00009        0        0        0        0        0        0
ERCC-00012        0        0        0        0        0        0
ERCC-00013        0        0        0        0        0        0
ERCC-00014        0        0        0        0        0        0
ERCC-00016        0        0        0        0        0        0
ERCC-00017        2        0        0        0        1        0
ERCC-00019        0        0        0        0        0        0</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="2" style="list-style-type: decimal"><li>Let’s check for duplicate sampleIDs and transcript IDs. Provided
there are no duplicate row or column names, the following should return
<code>interger(0)</code>.</li>
</ol><div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which</span><span class="op">(</span><span class="fu">duplicated</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">integer</span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which</span><span class="op">(</span><span class="fu">duplicated</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fu">integer</span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="3" style="list-style-type: decimal"><li>We’ll double check the sampleIDs match the sample information
file.</li>
</ol><div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu">identical</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">sampleID</span><span class="op">)</span><span class="op">)</span><span class="op">{</span><span class="kw">stop</span><span class="op">(</span><span class="op">)</span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): </code></pre>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4:<a class="anchor" aria-label="anchor" href="#challenge-4"></a>
</h3>
<div class="callout-content">
<p>Why did we get an error here?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>We renamed the samples in the sample information to remove spaces, so
we need to do the same here.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span>x <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"\ "</span>, replacement <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="4" style="list-style-type: decimal"><li>Finally, we’ll check that there are no missing values in the counts
matrix. Note, there will be many zeros, but we are ensuring that there
are no <code>NA</code> values or black records. The following should
return <code>FALSE</code>.</li>
</ol><div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allMissValues</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu">all</span><span class="op">(</span><span class="fu">is.na</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">|</span> <span class="va">x</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span><span class="op">}</span></span>
<span></span>
<span><span class="fu">allMissValues</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE</code></pre>
</div>
<p><br></p>
<p>Take a final look at the cleaned up matrix.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6
ERCC-00002        0        0        0        0        0        0
ERCC-00003        0        0        0        0        0        0
ERCC-00004        0        0        0        0        3        0
ERCC-00009        0        0        0        0        0        0
ERCC-00012        0        0        0        0        0        0
ERCC-00013        0        0        0        0        0        0
ERCC-00014        0        0        0        0        0        0
ERCC-00016        0        0        0        0        0        0
ERCC-00017        2        0        0        0        1        0
ERCC-00019        0        0        0        0        0        0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i rows, corresponding to the transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 22750 rows, corresponding to the transcript IDs"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i columns, corresponding to the samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">counts.mat.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 590 columns, corresponding to the samples"</code></pre>
</div>
<p><br></p>
<p>And save the output as a file.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">counts.mat.ibd</span>, file<span class="op">=</span><span class="st">"./data/counts.mat.ibd.txt"</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div id="document-your-reformatting-steps" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="document-your-reformatting-steps" class="callout-inner">
<h3 class="callout-title">Document your reformatting steps<a class="anchor" aria-label="anchor" href="#document-your-reformatting-steps"></a>
</h3>
<div class="callout-content">
<p>It is crucial that you document your data reformatting so that it is
reproducible by other researchers. Create a project folder for your data
processing and include:</p>
<ul><li>Raw input data files collected from ArrayExpress, GEO or
similar</li>
<li>The output counts matrix and target variable vector, stored in a
common format such at .txt format</li>
<li>The full R scripts used to perform all the data clean up steps.
Store the script with the input and output data.</li>
<li>A readme file that explains the steps taken, and how to run the
script on the inputs to generate the outputs. This should also include
details of all software versions used (e.g. R version, RStudio or
Jupyter Notebooks version, and the versions of any code libraries
used)</li>
</ul></div>
</div>
</div>
</section><section id="the-tb-dataset-another-example"><h2 class="section-heading">The TB Dataset: Another Example<a class="anchor" aria-label="anchor" href="#the-tb-dataset-another-example"></a>
</h2>
<hr class="half-width"><div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Take a look at <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-6845" class="external-link">The
Tuburculosis (TB) Dataset - E-MTAB-6845</a>. You can download the sdrf
file from zenodo and save it to your <code>data</code> directory by
running the following code.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8125141/files/E-MTAB-6845.sdrf.txt"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/E-MTAB-6845.sdrf.txt"</span><span class="op">)</span></span></code></pre>
</div>
<p>Read the sdrf file into R and take a look at the data. Make a list of
the potential issues with this data based on the checklist above? What
steps would need to be taken to get them ready to train a machine
learning classifier?</p>
<p>As a help, the code to read the file in from your data directory
is:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.tb</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-6845.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>

<div id="challenge-extension" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-extension" class="callout-inner">
<h3 class="callout-title">Challenge Extension<a class="anchor" aria-label="anchor" href="#challenge-extension"></a>
</h3>
<div class="callout-content">
<p>Can you correctly reorder the following code snippets to create a
pipeline to reformat the data in line with the sample information
checklist? Use the
<code>Characteristics[progressor status (median follow-up 1.9 years)]</code>
as the target variable.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span></span>
<span>  <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Comment[sample id]'</span>,</span>
<span>  <span class="st">'prog_status'</span> <span class="op">=</span> <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span></span>
<span>            <span class="st">'Comment[sample id]'</span>,</span>
<span>            <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span>  <span class="op">%&gt;%</span></span>
<span></span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">prog_status</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"non_progressor"</span>,<span class="st">"tb_progressor"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             class <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">class</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>   </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">recode</span><span class="op">(</span><span class="va">prog_status</span>,                                </span>
<span>              <span class="st">"non-progressor"</span> <span class="op">=</span> <span class="st">"non_progressor"</span>,</span>
<span>              <span class="st">"TB progressor"</span> <span class="op">=</span> <span class="st">"tb_progressor"</span><span class="op">)</span>, </span>
<span>              class <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">if_else</span><span class="op">(</span><span class="va">prog_status</span> <span class="op">==</span> <span class="st">'non_progressor'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>                 </span>
<span>              <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span></span>
<span><span class="va">samp.info.tb</span> <span class="op">%&gt;%</span></span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="op">)</span>     </span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Example code to clean up the sample information
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.tb</span> <span class="op">%&gt;%</span> </span>
<span>  </span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span></span>
<span>            <span class="st">'Comment[sample id]'</span>,</span>
<span>            <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span>  <span class="op">%&gt;%</span>             <span class="co"># select columns</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span></span>
<span>        <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Comment[sample id]'</span>,</span>
<span>        <span class="st">'prog_status'</span> <span class="op">=</span> <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># rename them</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>                                                                  <span class="co"># remove duplicate rows</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">recode</span><span class="op">(</span><span class="va">prog_status</span>,                                 <span class="co"># recode target variable</span></span>
<span>                    <span class="st">"non-progressor"</span> <span class="op">=</span> <span class="st">"non_progressor"</span>,</span>
<span>                    <span class="st">"TB progressor"</span> <span class="op">=</span> <span class="st">"tb_progressor"</span><span class="op">)</span>, </span>
<span>                    class <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">if_else</span><span class="op">(</span><span class="va">prog_status</span> <span class="op">==</span> <span class="st">'non_progressor'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>                 <span class="co"># create numerical class</span></span>
<span>                    <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span></span>
<span>       <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">prog_status</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"non_progressor"</span>,<span class="st">"tb_progressor"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                     class <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">class</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>                          <span class="co"># set variables to factors</span></span>
<span></span>
<span>       <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="op">)</span>                                                                     <span class="co"># view output</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 360
Columns: 3
$ sampleID    &lt;chr&gt; "PR123_S19", "PR096_S13", "PR146_S14", "PR158_S12", "PR095…
$ prog_status &lt;fct&gt; non_progressor, non_progressor, non_progressor, non_progre…
$ class       &lt;fct&gt; -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1…</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Machine learning algorithms require specific data input formats, and
for data to be consistently formatted across variables and samples. A
classification task for example requires a matrix of the value of each
input variable for each sample, and the class label for each
sample.</li>
<li>Clinical sample information is often messy, with inconsistent
formatting as a result of how it is collected. This applies to data
downloaded from public repositories.</li>
<li>You must carefully check all data for formatting and data integrity
issues that may negatively impact your downstream ml analysis.</li>
<li>Document your data reformatting and store the code pipeline used
along with the raw and reformatted data to ensure your procedure is
reproducible</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="episode3.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="episode5.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="episode3.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Data Collection: GEO
        </a>
        <a class="chapter-link float-end" href="episode5.html" rel="next">
          Next: Data Readiness:... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode4.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:parkinsone@cardiff.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.14.1" class="external-link">sandpaper (0.14.1)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.1" class="external-link">pegboard (0.7.1)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.3.3" class="external-link">varnish (0.3.3)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode4.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "machine learning, RNA-Seq, data readiness, pre-processing, data management",
  "name": "Data Readiness: Data Format and Integrity",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode4.html",
  "identifier": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/episode4.html",
  "dateCreated": "2023-05-23",
  "dateModified": "2023-11-10",
  "datePublished": "2023-11-10"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

