<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>RNA-Seq: Data Readiness for Machine Learning Applications: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="manifest" href="../site.webmanifest">
<link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      RNA-Seq: Data Readiness for Machine Learning Applications
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            RNA-Seq: Data Readiness for Machine Learning Applications
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-Seq: Data Readiness for Machine Learning Applications
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="episode1.html">1. Introduction: Machine Learning Ready RNA-Seq Data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="episode2.html">2. Data Collection: ArrayExpress</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="episode3.html">3. Data Collection: GEO</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="episode4.html">4. Data Readiness: Data Format and Integrity</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="episode5.html">5. Data Readiness: Technical Noise</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="episode6.html">6. Data Readiness: Distribution and Scale</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-episode1"><p>Content from <a href="episode1.html">Introduction: Machine Learning Ready RNA-Seq Data</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode1.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 11 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What characteristics of a dataset do I need to consider to make it
‘ready’ for a machine learning /AI modelling analysis?</li>
<li>Where can I find a publicly available RNA-Seq dataset suitable for a
machine learning analysis?</li>
<li>What format is RNA-Seq data stored in on public repositories?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the factors that must be considered in readying an RNA-Seq
dataset for a machine learning analysis?</li>
<li>Recall the main platforms hosting functional genomic datasets</li>
<li>Explain the difference between raw and processed RNA-Seq data, and
the different stages of processing for RNA-Seq data</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="dataset-characteristics-for-supervised-learning"><h2 class="section-heading">Dataset Characteristics for supervised learning<a class="anchor" aria-label="anchor" href="#dataset-characteristics-for-supervised-learning"></a>
</h2>
<hr class="half-width">
<p>The success of a ML/AI model depends heavily on the input data.
Finding an appropriate dataset can be a challenge and the data must be
selected carefully as a predictive model trained on the unreliable or
inappropriate data will produce unreliable predictions.</p>
<p>There is currently no curated “ML/AI ready” datasets that meet the
requirements of machine learning analyses within public functional
genomics repositories. It is therefore important to examine a data set
carefully to confirm its appropriateness for a supervised machine
learning task. The following characteristics of the data should be
considered:</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="75%">
</colgroup>
<thead><tr class="header">
<th>Characteristic</th>
<th>Considerations</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Quality and Provenance</strong></td>
<td>Assess source of the data, its likely quality and whether it is
recognised by the community, for example having been used in highly
regarded publications. Datasets must have sufficient metadata, including
accurate and reliable class labels, ideally connected to a
domain-specific or community-specific ontology so that the meaning is
widely interpretable and comparable between studies.</td>
</tr>
<tr class="even">
<td><strong>Size</strong></td>
<td>Ensure the data set captures the full complexity of the underlying
distribution. If there is significant biological heterogeneity in the
sample population, will there be sufficient samples to create
independent and representative train, validation and tests sets? Bulk
RNA-Seq datasets with only a few samples for each group are unlikely to
be sufficient to train a machine learning classifier.</td>
</tr>
<tr class="odd">
<td><strong>Format and Integrity</strong></td>
<td>Machine learning models require data to be machine readable and
input in a specific format. Data collected experimentally and data
acquired from public repositories will likely need to be reformatted and
checked for integrity issues.</td>
</tr>
<tr class="even">
<td><strong>Technical Noise</strong></td>
<td>RNA-Seq data is likely to contain some technical noise resulting
from experiential processes that is unrelated to the underlying biology.
This has the potential to bias machine learning algorithms and should be
identified and ideally removed.</td>
</tr>
<tr class="odd">
<td><strong>Distribution and Scale</strong></td>
<td>Many machine learning algorithms are sensitive to the distribution
and the scale of the input data. RNA-Seq data must be transformed and
re-scaled to perform well in a machine learning analysis.</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>In this lesson, we will address each of these issues in turn, working
through the collection of a RNA-Seq dataset, reformatting, integrity
checking, noise elimination, transformation and rescaling to make the
data ready for machine learning. Further information on some of the data
considerations in supervised machine learning with biological data can
be found in the <a href="https://www.nature.com/articles/s41592-021-01205-4" class="external-link">DOME
recommendations</a>.</p>
</section><section id="functional-genomics-platforms"><h2 class="section-heading">Functional Genomics platforms<a class="anchor" aria-label="anchor" href="#functional-genomics-platforms"></a>
</h2>
<hr class="half-width">
<p>There are two major public repositories for sourcing public
functional genomics data sets, in particular microarray and
next-generation sequencing data. Here we focus on NGS data, in
particular RNA-Seq transcriptomics datasets:</p>
<ol style="list-style-type: decimal">
<li><p>The <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies" class="external-link">Array
Express</a> collection within the Bio Studies database, maintained by
the European Molecular Biology Laboratory - European Bioinformatics
Institute (EMBL-EBI).</p></li>
<li><p>The <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression
Onmibus (GEO)</a>, maintained by the National Center for Biotechnology
Information, part of the US National Institues of Health.</p></li>
</ol>
<p>Each database contains thousands of RNA-Seq datasets relating to a
wide variety of experiments, not all of which would be suitable for
analysis with a machine learning approach.</p>
</section><section id="raw-vs.-processed-data"><h2 class="section-heading">Raw vs. Processed Data<a class="anchor" aria-label="anchor" href="#raw-vs.-processed-data"></a>
</h2>
<hr class="half-width">
<p>Public datasets will contain both raw and processed RNA-Seq data, as
well as information about the relationship between the RNA-Seq data and
study samples, and the design of the experiment. Here is a brief summary
of the difference between raw and processed RNA-Seq data:</p>
<p><br></p>
<table style="width:100%;" class="table">
<colgroup>
<col width="16%">
<col width="66%">
<col width="16%">
</colgroup>
<thead><tr class="header">
<th>Data Type</th>
<th>What it is</th>
<th>File Format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Raw data</strong></td>
<td>Sequences of each read obtained from the sequencing instrument,
along with quality scores for each sequence</td>
<td>FASTQ</td>
</tr>
<tr class="even">
<td><strong>Processed data</strong></td>
<td>Matrix of abundance values (e.g. integer counts) for the feature of
interest, which may be genes, transcripts, exons or miRNAs</td>
<td>.csv / .txt / .tsv</td>
</tr>
</tbody>
</table>
<p>For more information on the fastq file format, see <a href="https://www.youtube.com/watch?v=tO2H3zuBouw" class="external-link">RDMBites | FASTQ
Format</a> and for information on the processing of raw fastq to gene
abundance data, see <a href="https://www.youtube.com/watch?v=3Pe9xcGF_Wo" class="external-link">RDMbites | RNAseq
expression data</a>.</p>
<p><br></p>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>The description “processed” means different things in different
studies. Things to be aware of:</p>
<ol style="list-style-type: decimal">
<li>Data is often generated and made available at different levels:</li>
</ol>
<ul>
<li>
<strong>Raw abundance</strong> of sequencing reads for each feature
of interest (e.g. integer counts of the number of reads mapped to a
given gene)</li>
<li>
<strong>Normalised abundance </strong>, further processed to account
for sequence depth and/or gene length such as TPM and FPKM (output using
DESeq2, edgeR)</li>
<li>
<strong>Normalised and transformed</strong> adundance with
normalisation and further transformation applied such as log2 or
variance stabilising transformation (vst) applied</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><p>The features of interest, say transcripts, may also have been
filtered to remove transcripts with read counts below a particular
threshold.</p></li>
<li><p>Transcripts may have been summarised to genes.</p></li>
<li><p>The data may be given as a matrix of features of interest and
samples. Alternatively there may be a separate file for each sample that
need to be combined.</p></li>
</ol>
<p><strong>It is important to carefully read the protocols to understand
determine how the data has been processed and if is appropriate for your
task.</strong></p>
</div>
</div>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>What type of data do you need to look for in a gene expression data
repository as input to a machine learning modelling analysis?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>The starting point for a machine learning model is processed RNA-Seq
data, in the form of count data. This could be raw count data, or TPM or
FPKM pre-processed data.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Sourcing and appropriate RNA-Seq dataset and preparing it for a
machine learning analysis requires you to consider the dataset quality,
size, format, noise content, data distribution and scale</li>
<li>The two main repositories for sourcing RNA-Seq datasets are
ArrayExpress and GEO</li>
<li>Processed data, in the form of raw counts, or further processed
counts data is the starting point for machine learning analysis</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-episode2"><p>Content from <a href="episode2.html">Data Collection: ArrayExpress</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode2.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 15 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What format are processed RNA-Seq dataset store on Array
Express?</li>
<li>How do I search for a dataset that meets my requirements on Array
Express?</li>
<li>Which files do I need to download?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the expected data and file formats for processed RNA-Seq
datasets on ArrayExpress</li>
<li>Demonstrate ability to correctly identify the subset of files
required to create a dataset for a simple supervised machine learning
task (e.g., binary classification)</li>
<li>Demonstrate ability to download and save processed RNA-Seq data
files from ArrayExpress suitable for a simple supervised machine
learning task (e.g., binary classification)</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="arrayexpress-file-formats"><h2 class="section-heading">ArrayExpress File Formats<a class="anchor" aria-label="anchor" href="#arrayexpress-file-formats"></a>
</h2>
<hr class="half-width">
<p>Array Express data sets comprise of four important file types that
contain all of the information about an experiment:</p>
<p><br></p>
<table class="table">
<colgroup>
<col width="25%">
<col width="75%">
</colgroup>
<thead><tr class="header">
<th>File Type</th>
<th>What It Is</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Sample and Data Relationship Format (SDRF)</strong></td>
<td>Tab-delimited file containing information about the samples, such as
phenotypical information, and factors that may be of interest in a
classification task. This file may also contain information about the
relationships between the samples, and details of the file names of the
raw data fastq for each sample.</td>
</tr>
<tr class="even">
<td><strong>Investigation Description Format (IDF)</strong></td>
<td>Tab-delimited file containing top level information about the
experiment including title, description, submitter contact details and
protocols.</td>
</tr>
<tr class="odd">
<td><strong>Raw data files</strong></td>
<td>Sequence data and quality scores, typically stored in FASTQ format.
There is however inconsistency - in some data sets, the term raw data
refers to processed data in the form of raw counts data, meaning RNA-Seq
counts that have not undergone further processing.</td>
</tr>
<tr class="even">
<td><strong>Processed data files</strong></td>
<td>Processed data which includes raw counts or further processed
abundance measures such as TPM and FPKM. The data may be provided as a
separate file for each sample, or as a matrix of all samples</td>
</tr>
</tbody>
</table>
<p>Detailed information about BioStudies file formats is available in
the <a href="https://www.ebi.ac.uk/biostudies/misc/MAGE-TABv1.1_2011_07_28.pdf" class="external-link">MAGE-TAB
Specification document</a>.</p>
<p><br></p>
</section><section id="searching-for-a-dataset-on-arrayexpress"><h2 class="section-heading">Searching for a Dataset on ArrayExpress<a class="anchor" aria-label="anchor" href="#searching-for-a-dataset-on-arrayexpress"></a>
</h2>
<hr class="half-width">
<p>We will use ArrayExpress to select a RNA-Seq dataset with a case
control design, suitable for constructing a machine learning
classification model. Machine learning based analyses generally perform
better the larger the sample size, and may perform very poorly and give
misleading results with insufficient samples, so we’ll look for a
datasets with a relatively large number of samples.</p>
<p>Let’s begin on the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies" class="external-link">ArrayExpress
home page</a>.</p>
<p><br></p>
<figure><img src="../fig/biostudies_home_page.jpg" alt="Screen Shot of Array Express Collection Tab" class="figure mx-auto d-block"><figcaption>ArrayExpress Home Page</figcaption></figure><p><br></p>
<p>We’ll apply the following filters to refine the results set, leaving
all other filters as the default with no selection:</p>
<p><br></p>
<table class="table">
<thead><tr class="header">
<th>Filter</th>
<th>Selection</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Study type</td>
<td>rna-seq of coding rna</td>
</tr>
<tr class="even">
<td>Organism</td>
<td>homo sapiens</td>
</tr>
<tr class="odd">
<td>Technology</td>
<td>sequencing assay</td>
</tr>
<tr class="even">
<td>Assay by Molecule</td>
<td>rna assay</td>
</tr>
<tr class="odd">
<td>Processed Data Available</td>
<td>yes (check box)</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>Now use the sort in the top right of the screen to sort the results
by “Samples”, in descending order. The results should look similar to
this:</p>
<p><br></p>
<figure><img src="../fig/biostudies_home_filtered.jpg" alt="Screen Shot of Array Express Collection Tab Filtered Results" class="figure mx-auto d-block"><figcaption>ArrayExpress Filtered Results</figcaption></figure></section><section id="illustrative-dataset-ibd-dataset"><h2 class="section-heading">Illustrative Dataset: IBD Dataset<a class="anchor" aria-label="anchor" href="#illustrative-dataset-ibd-dataset"></a>
</h2>
<hr class="half-width">
<p>In the search results displayed above, the top ranked dataset is
E-MTAB-11349: <em>Whole blood expression profiling of patients with
inflammatory bowel diseases in the IBD-Character cohort</em>. We’ll call
it the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-11349" class="external-link">IBD
dataset</a>.</p>
<p>The IBD dataset comprises human samples of patients with inflammatory
bowel diseases and controls. The ‘Protocols’ section explains the main
steps in the generation of the data. The nucleic acid sequencing
protocol gives details of the sequencing platform used to generate the
raw fastq data, and the version of the human genome used in the
alignment step to generate the count data. The normalization data
transformation protocol gives the tools used to normalise raw counts
data for sequence depth and sample composition, in this case
normalisation is conducted using the R package DESeq2. Let’s look at
some of the basic information on this dataset:</p>
<p><br></p>
<table class="table">
<thead><tr class="header">
<th>Data Field</th>
<th>Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Sample count</td>
<td>590</td>
</tr>
<tr class="even">
<td>Experimental Design</td>
<td>case control design</td>
</tr>
<tr class="odd">
<td>Experimental factors</td>
<td>Normal, Ulcerative colitis, Crohn’s disease</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>The dataset contains two alternative sets of processed data, a matrix
of raw counts, and a matrix of processed counts that have been
normalised using DESeq2, and filtered to only include transcripts with a
mean raw expression read count &gt; 10.</p>
</section><section id="downloading-and-reading-into-r"><h2 class="section-heading">Downloading and Reading into R<a class="anchor" aria-label="anchor" href="#downloading-and-reading-into-r"></a>
</h2>
<hr class="half-width">
<p>Let’s download the SDRF file and the raw counts matrix - these are
the two files that contain the information we will need to build a
machine learning classification model. In the data files box to the
right hand side, check the follwoing two files
<code>E-MTAB-11349.sdrf.txt</code> and
<code>ArrayExpress-raw.csv</code>, and save in a folder called
<code>data</code> in the working directory for your R workbook. For
consistency, rename <code>ArrayExpress-raw.csv</code> as
<code>E-MTAB-11349.counts.matrix.csv</code>.</p>
<p><br></p>
<div class="section level3">
<h3 id="raw-counts-matrix">Raw counts matrix<a class="anchor" aria-label="anchor" href="#raw-counts-matrix"></a>
</h3>
<p>In R studio, open your project workbook and read in the raw counts
data. Then check the dimensions of the matrix to confirm we have the
expected number of samples and transcript IDs.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/E-MTAB-11349.counts.matrix.csv"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">","</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in file(file, "rt"): cannot open file
'data/E-MTAB-11349.counts.matrix.csv': No such file or directory</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in file(file, "rt"): cannot open the connection</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">writeLines</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i %s"</span>, <span class="fu">c</span><span class="op">(</span><span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"rows corresponding to transcript IDs"</span>, <span class="st">"columns corresponding to samples"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'raw.counts.ibd' not found</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>View a small subset of the data, (e.g. first ten rows and 8 columns)
to see how the samples are labelled and how the sample and transcript
IDs are formatted.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'raw.counts.ibd' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sdrf-file">SDRF File<a class="anchor" aria-label="anchor" href="#sdrf-file"></a>
</h3>
<p>Now let’s read the sdrf file into R and check the dimensions of the
file.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in the sdrf file</span></span>
<span></span>
<span><span class="va">samp.info.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/E-MTAB-11349.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i rows, corresponding to the samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 590 rows, corresponding to the samples"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i columns, corresponding to the available variables for each sample"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "There are 32 columns, corresponding to the available variables for each sample"</code></pre>
</div>
<p>If we view the column names, we can see that the file does indeed
contain a set of variables describing both phenotypical and experimental
protocol information relating to each sample.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "Source Name"                            
 [2] "Characteristics[organism]"              
 [3] "Characteristics[age]"                   
 [4] "Unit[time unit]"                        
 [5] "Term Source REF"                        
 [6] "Term Accession Number"                  
 [7] "Characteristics[developmental stage]"   
 [8] "Characteristics[sex]"                   
 [9] "Characteristics[individual]"            
[10] "Characteristics[disease]"               
[11] "Characteristics[organism part]"         
[12] "Material Type"                          
[13] "Protocol REF"                           
[14] "Protocol REF"                           
[15] "Protocol REF"                           
[16] "Extract Name"                           
[17] "Material Type"                          
[18] "Comment[LIBRARY_LAYOUT]"                
[19] "Comment[LIBRARY_SELECTION]"             
[20] "Comment[LIBRARY_SOURCE]"                
[21] "Comment[LIBRARY_STRATEGY]"              
[22] "Protocol REF"                           
[23] "Performer"                              
[24] "Assay Name"                             
[25] "Technology Type"                        
[26] "Protocol REF"                           
[27] "Derived Array Data File"                
[28] "Comment [Derived ArrayExpress FTP file]"
[29] "Protocol REF"                           
[30] "Derived Array Data File"                
[31] "Comment [Derived ArrayExpress FTP file]"
[32] "Factor Value[disease]"                  </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>ArrayExpress stores two standard files with information about each
experiment: 1. Sample and Data Relationship Format (SDRF) and 2.
Investigation Description Format (IDF).</li>
<li>ArrayExpress provides raw and processed data for RNA-Seq datasets,
typically stored as csv, tsv, or txt files.</li>
<li>The filters on the ArrayExpress website allow you to select and
download a dataset that suit your task.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-episode3"><p>Content from <a href="episode3.html">Data Collection: GEO</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode3.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 20 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What format are processed RNA-Seq dataset store on GEO?</li>
<li>How do I search for a dataset that meets my requirements on
GEO?</li>
<li>How do I use the R package <code>GEOquery</code> to download
datasets from GEO into R?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the expected data and file formats for processed RNA-Seq
datasets on GEO</li>
<li>Demonstrate ability to correctly identify the subset of files
required to create a dataset for a simple supervised machine learning
task (e.g., binary classification)</li>
<li>Demonstrate ability to download and save processed RNA-Seq data
files from GEO suitable for a simple supervised machine learning task
(e.g., binary classification)</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="geo-data-objects"><h2 class="section-heading">GEO Data Objects<a class="anchor" aria-label="anchor" href="#geo-data-objects"></a>
</h2>
<hr class="half-width">
<p>NCBI ‘gene expression omnibus’ , GEO, one of the largest repositories
for functional genomics data. GEO data are stored as individual samples,
series of samples and curated datasets. Accession numbers of GEO broadly
fall into four categories:</p>
<ul>
<li>GSM - GEO Sample. These refer to a single sample.</li>
<li>GSE - GEO Series. These refer to a list of GEO samples that together
form a single experiment.</li>
<li>GDS - GEO Dataset. These are curated datasets containing summarised
information from the underlying GEO samples.</li>
</ul>
<p>There are many more datasets available as series than there are
curated GEO Datasets. For this lesson, we’ll focus finding a GEO Series
as an illustrative dataset.</p>
</section><section id="geo-file-formats"><h2 class="section-heading">GEO File Formats<a class="anchor" aria-label="anchor" href="#geo-file-formats"></a>
</h2>
<hr class="half-width">
<p>Data for RNA-Seq studies have three components on GEO (similar to
microarray studies):</p>
<ol style="list-style-type: decimal">
<li><strong>Metadata</strong></li>
</ol>
<p>The metadata is equivalent to sample data relationship information
provided on ArrayExpress and refers to the descriptive information about
the overall study and individual samples, as well as protocol
information and references to the raw and processed data files names for
each sample. Metadata is typically stored in a number of standard
formats on GEO:</p>
<ul>
<li>SOFT (Simple Omnibus Format in Text) is a line-based plain text
format</li>
<li>MINiML is an alternative to SOFT and offers an XML rendering of
SOFT</li>
<li>Series Matrix is a summary text files that include a tab delimited
value-matrix table</li>
</ul>
<p>The raw and processed counts data is typically not included in any of
the above files for an RNA-Seq experiement (whereas they may be included
with a microarray experiment).</p>
<ol start="2" style="list-style-type: decimal">
<li><strong>Processed data files</strong></li>
</ol>
<p>Processed data (raw counts, noramlised counts) are typically provided
in a supplementary file, often a gz archieves of text files. Raw and
normalised counts may be provided as separate supplementary files.</p>
<ol start="3" style="list-style-type: decimal">
<li><strong>Raw data files</strong></li>
</ol>
<p>Raw data files provide the original reads and quality scores and are
stored in the NIH’s <a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">Sequence
Read Archieve (SRA)</a>.</p>
<p>Further information on how RNA-Seq experiments are stored on GEO is
given in the <a href="https://www.ncbi.nlm.nih.gov/geo/info/seq.html" class="external-link">submission
guidelines</a></p>
</section><section id="searching-for-a-dataset-on-geo"><h2 class="section-heading">Searching for a Dataset on GEO<a class="anchor" aria-label="anchor" href="#searching-for-a-dataset-on-geo"></a>
</h2>
<hr class="half-width">
<p>Let’s begin on the <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO
home page</a>. GEO provides online search and filter tools. We search by
adding “rna-seq” as a keyword in the search box, and then apply the
following filters to the subsequent results page, and rank the results
by the number of samples descending:</p>
<p><br></p>
<table class="table">
<thead><tr class="header">
<th>Filter</th>
<th>Selection</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Keywords.</td>
<td>“rna-seq”</td>
</tr>
<tr class="even">
<td>Entry type</td>
<td>Series or DataSets</td>
</tr>
<tr class="odd">
<td>Organism</td>
<td>homo sapiens</td>
</tr>
<tr class="even">
<td>Sample count</td>
<td>From 100 to 1,000</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>The full search string is:</p>
<p><code>rna-seq[All Fields] AND "Homo sapiens"[Organism] AND (("gse"[Filter] OR "gds"[Filter]) AND ("100"[n_samples] : "1000"[n_samples]))</code></p>
<p><br></p>
<p>The filtered results page should looks like this:</p>
<figure><img src="../fig/geo_home_filtered.jpg" alt="Screen Shot of GEO DataSets Search Results Tab Filtered" class="figure mx-auto d-block"><figcaption>GEO Filtered Results</figcaption></figure></section><section id="illustrative-dataset-covid-19-dataset"><h2 class="section-heading">Illustrative Dataset: Covid-19 Dataset<a class="anchor" aria-label="anchor" href="#illustrative-dataset-covid-19-dataset"></a>
</h2>
<hr class="half-width">
<p>The data set GSE212041 relates to the case-control study of
neutrophils in Covid-19+ patients named “Longitudinal characterization
of circulating neutrophils uncovers phenotypes associated with severity
in hospitalized COVID-19 patients” and comprises longitudinal human
samples from 306 hospitalized COVID-19+ patients, 78 symptomatic
controls, and 8 healthy controls, each at multiple time points, with a
total of 781 samples. We’ll call it the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE212041" class="external-link">Covid-19
dataset</a>. Processed data is provided as raw counts and TPM normalised
counts in the Supplementary files.</p>
<p>Let’s look at some of the basic information on this dataset:</p>
<p><br></p>
<table class="table">
<thead><tr class="header">
<th>Data Field</th>
<th>Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Sample count</td>
<td>781</td>
</tr>
<tr class="even">
<td>Experimental Design</td>
<td>case control design</td>
</tr>
<tr class="odd">
<td>Experimental factors</td>
<td>COVID+, COVID- symptomatic, healthy</td>
</tr>
</tbody>
</table></section><section id="downloading-and-reading-into-r"><h2 class="section-heading">Downloading and Reading into R<a class="anchor" aria-label="anchor" href="#downloading-and-reading-into-r"></a>
</h2>
<hr class="half-width">
<p><br></p>
<div class="section level3">
<h3 id="using-geoquery-to-download-a-geo-series-as-an-expression-set">Using GEOquery to download a GEO series as an expression set<a class="anchor" aria-label="anchor" href="#using-geoquery-to-download-a-geo-series-as-an-expression-set"></a>
</h3>
<p>The function <code>getGEO()</code> from the <code>GEOquery</code>
library provides a convenient way to download GEO SOFT format data into
R. Here we will download the GEO Series GSE212041. This blog provides
additional information on <a href="https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/" class="external-link">reading
GEO SOFT files into R</a></p>
<p>If there is more than one SOFT file for a GEO Series,
<code>getGEO()</code> will return a list of datasets. Let’s download
GSE212041.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gse212041</span> <span class="op">&lt;-</span> <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEO</span><span class="op">(</span><span class="st">"GSE212041"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Setting options('download.file.method.GEOquery'='auto')</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Setting options('GEOquery.inmemory.gpl'=FALSE)</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Found 2 file(s)</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GSE212041</span><span class="op">-</span><span class="va">GPL18573_series_matrix.txt.gz</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GSE212041</span><span class="op">-</span><span class="va">GPL24676_series_matrix.txt.gz</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Number of files downloaded: %i"</span>, <span class="fu">length</span><span class="op">(</span><span class="va">gse212041</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Number of files downloaded: 2"</code></pre>
</div>
<div id="challenge-1-how-many-samples-are-in-each-file" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-how-many-samples-are-in-each-file" class="callout-inner">
<h3 class="callout-title">Challenge 1: How many samples are in each
file?<a class="anchor" aria-label="anchor" href="#challenge-1-how-many-samples-are-in-each-file"></a>
</h3>
<div class="callout-content">
<p>Write the code to check that the number of samples in each file gives
us the total expected number of 781 samples.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">writeLines</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"file %i: %i samples"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fu">c</span><span class="op">(</span><span class="fu">dim</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>file 1: 16 samples
file 2: 765 samples</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<p>Extract the metadata for the larger dataset and examine the column
names to verify that the file contains the expected metadata about the
experiment.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.cov19</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.cov19</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "title"                   "geo_accession"          
 [3] "status"                  "submission_date"        
 [5] "last_update_date"        "type"                   
 [7] "channel_count"           "source_name_ch1"        
 [9] "organism_ch1"            "characteristics_ch1"    
[11] "characteristics_ch1.1"   "characteristics_ch1.2"  
[13] "characteristics_ch1.3"   "characteristics_ch1.4"  
[15] "molecule_ch1"            "extract_protocol_ch1"   
[17] "extract_protocol_ch1.1"  "taxid_ch1"              
[19] "description"             "description.1"          
[21] "data_processing"         "data_processing.1"      
[23] "data_processing.2"       "data_processing.3"      
[25] "data_processing.4"       "platform_id"            
[27] "contact_name"            "contact_email"          
[29] "contact_phone"           "contact_laboratory"     
[31] "contact_institute"       "contact_address"        
[33] "contact_city"            "contact_state"          
[35] "contact_zip/postal_code" "contact_country"        
[37] "data_row_count"          "instrument_model"       
[39] "library_selection"       "library_source"         
[41] "library_strategy"        "supplementary_file_1"   
[43] "acuity.max:ch1"          "cell type:ch1"          
[45] "covid-19 status:ch1"     "patient category:ch1"   
[47] "time point:ch1"         </code></pre>
</div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="downloading-supplementary-files">Downloading supplementary files<a class="anchor" aria-label="anchor" href="#downloading-supplementary-files"></a>
</h3>
<p>Notice in this series, the counts matrix is not contained in the
exprs() slot of the expression set. The expression set only contains a
list of the accession numbers of the samples included in the expression
set, but not the actual count data. We can verify this by looking at the
dimensions of the object in the exprs slot.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu">exprs</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]   0 765</code></pre>
</div>
<p>To get the counts data, we need to download the Supplementary file
(not downloaded by the <code>getGeo()</code> function). The
<code>GEOquery::getGEOSuppFiles()</code> function enables us to view and
download supplementary files attached to a GEO Series (GSE). This
function doesn’t parse the downloaded files, since the file format may
vary between data series.</p>
<p>Set the ‘fetch_files’ argument to FALSE initially to view the
available files.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE212041"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                          <span class="op">)</span><span class="op">$</span><span class="va">fname</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
[1] "GSE212041_Neutrophil_RNAseq_Count_Matrix.txt.gz" "GSE212041_Neutrophil_RNAseq_TPM_Matrix.txt.gz" 
</code></pre>
</div>
<div id="challenge-2-download-the-dataset" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2-download-the-dataset" class="callout-inner">
<h3 class="callout-title">Challenge 2: Download the dataset<a class="anchor" aria-label="anchor" href="#challenge-2-download-the-dataset"></a>
</h3>
<div class="callout-content">
<p>Now Download the file to your /data subdirectory, by setting
<code>fetch_files</code> to TRUE and apply a <code>filter_regex</code>
of <code>'*Count_Matrix*'</code> to include the raw counts file only.
Read a sample of the file into R and check the dimensions of the file
correspond to the expected number of samples.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE212041"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="st">'*Count_Matrix*'</span></span>
<span>                          <span class="op">)</span> </span>
<span></span>
<span><span class="va">raw.counts.cov19</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/GSE212041_Neutrophil_RNAseq_Count_Matrix.txt.gz"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">"\t"</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i rows, corresponding to transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.cov19</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i columns, corresponding to samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.cov19</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
[1] "60640 rows, corresponding to transcript IDs"
[1] "783 columns, corresponding to samples"
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-3-download-dataset-gse157657-from-geo" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3-download-dataset-gse157657-from-geo" class="callout-inner">
<h3 class="callout-title">Challenge 3: Download dataset GSE157657 from
GEO<a class="anchor" aria-label="anchor" href="#challenge-3-download-dataset-gse157657-from-geo"></a>
</h3>
<div class="callout-content">
<p>Use the functions from the <code>GEOquery</code> library to download
the counts data and sample information for <code>GSE157657</code>, a
dataset of 760 TB patients.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gse157657</span> <span class="op">&lt;-</span> <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEO</span><span class="op">(</span><span class="st">"GSE157657"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samp.info.geotb</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">gse157657</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE157657"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                          <span class="op">)</span> </span>
<span></span>
<span><span class="va">raw.counts.geotb</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/GSE157657_norm.data.txt.gz"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">"\t"</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i rows, corresponding to transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.geotb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i columns, corresponding to samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.geotb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "58735 rows, corresponding to transcript IDs"
[1] "761 columns, corresponding to samples"
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4:<a class="anchor" aria-label="anchor" href="#challenge-4"></a>
</h3>
<div class="callout-content">
<p>How have the counts data been processed in this dataset?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">
  Give me a hint
  </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>Hint: you’ll need to refer back to the GEO website to understand how
the data has been processed.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>The counts data have been normalised and transformed using a variance
stabilising transformation using the <code>DESeq2</code> R library.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Similar to ArrayExpress, GEO stores samples information and counts
matrices separately. Sample information is typically stored in
<code>SOFT</code> format, as well as a <code>.txt</code> file.</li>
<li>Counts data may be stored as raw counts and some further processed
form of counts, typically as supplementary files. Always review the
documentation to determine what “processed data” refers to for a
particular dataset.</li>
<li>
<code>GEOquery</code> provides a convenient way to download files
directly from GEO into R.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-episode4"><p>Content from <a href="episode4.html">Data Readiness: Data Format and Integrity</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode4.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 40 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What format do I require my data in to build a supervised machine
learning classifier?</li>
<li>What are the potential data format and data integrity issues I will
encounter with RNA-Seq datasets, including those downloaded from public
repositories, that I’ll need to address before beginning any
analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the format of a dataset required as input to a simple
supervised machine learning analysis</li>
<li>Recall a checklist of common formatting and data integrity issues
encountered with sample information (metadata) in real RNA-Seq datasets,
and why each checklist item presents a problem in downstream ML
analysis</li>
<li>Recall a checklist of common formatting and data integrity issues
encountered with counts matrix RNA-Seq datasets, and why each checklist
item presents a problem in downstream ML analysis</li>
<li>Apply the checklist to an unseen dataset downloaded from either the
Array Express or GEO platform to identify potential issues</li>
<li>Construct a reusable R code pipeline implement required changes to a
new dataset</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="required-format-for-machine-learning-libraries"><h2 class="section-heading">Required format for machine learning libraries<a class="anchor" aria-label="anchor" href="#required-format-for-machine-learning-libraries"></a>
</h2>
<hr class="half-width">
<p>Data must be correctly formatted for use in a machine learning / AI
pipeline. The garbage in garbage out principle applies; a machine
learning model is only as good as the input data. This episode discusses
the kind of data requesed as input to a machine learing classification
model, and illustrates the processes formatting an RNA-Seq dataset
downloaded from a public repository, highlighting some of the likely
data format and integrity issues that learners may encounter.</p>
<p>For a supervised machine learning classification task, we
require:</p>
<ol style="list-style-type: decimal">
<li><p><strong>A matrix of values of all of the predictor
variables</strong> to be included in a machine learning model for each
of the samples. Example predictor variables would be genes obtained from
the counts data. Additional predictor variables may be obtained from the
sample information including demographic data (e.g. <code>sex</code>,
<code>age</code>) and clinical and laboratory measures.</p></li>
<li><p><strong>A vector of values for the target variable</strong> that
we are trying to predict for each sample. Example target variables for a
classification model would be disease state (the analogy for a
regression task would be a continuous variable such as a disease
progression or severity score).</p></li>
</ol>
<p>We will follow a series of steps to reformat and clean up the sample
information and counts matrix file for the IBD dataset collected form
ArrayExpress by applying a checklist of data clean up items. We’ll start
by reading in the dataset again from the data subfolder.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-11349.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">raw.counts.ibd</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-11349.counts.matrix.csv"</span>, sep<span class="op">=</span><span class="st">","</span> , header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in file(file, "rt"): cannot open file
'./data/E-MTAB-11349.counts.matrix.csv': No such file or directory</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in file(file, "rt"): cannot open the connection</code></pre>
</div>
</section><section id="idb-dataset-sample-information-clean-up"><h2 class="section-heading">IDB dataset: Sample Information Clean Up<a class="anchor" aria-label="anchor" href="#idb-dataset-sample-information-clean-up"></a>
</h2>
<hr class="half-width">
<p>In real RNA-Seq experiments, sample information (metadata) has likely
been collected from a range of sources, for example medical records,
clinical and laboratory readings and then manually compiled in a matrix
format, often a spreadsheet, before being submitted to a public
respository. Each of these steps has the potential to introduce errors
and inconsistency in the data that must be investigated prior to any
downstream machine learning analysis. Whether you are working with
sample information received directly from co-workers or downloaded from
a public repsository, the following guidelines will help identify and
resolve issues that may impact downstream analysis.</p>
<div id="sample-information-readiness-checklist" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="sample-information-readiness-checklist" class="callout-inner">
<h3 class="callout-title">Sample Information Readiness Checklist<a class="anchor" aria-label="anchor" href="#sample-information-readiness-checklist"></a>
</h3>
<div class="callout-content">
<table class="table">
<colgroup>
<col width="7%">
<col width="42%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Item</th>
<th>Check For…</th>
<th>Rationale</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>
<strong>Unique identifier</strong> matching between counts matrix
and sample information</td>
<td>Target variables must be matched to predictor variables for each
sample</td>
</tr>
<tr class="even">
<td>2</td>
<td>
<strong>Appropriate target variable</strong> present in the
data</td>
<td>Sample information needs to contain the measure you are trying in
your machine learning model</td>
</tr>
<tr class="odd">
<td>3</td>
<td>
<strong>Unique variable names</strong>, without spaces or special
characters</td>
<td>Predictor variables may be drawn from the sample information. ML
algorithms require unique variables. Spaces etc. may cause errors with
bioinformatics and machine learning libraries used downstream</td>
</tr>
<tr class="even">
<td>4</td>
<td>
<strong>Machine readable and consistent encoding</strong> of
categorical variables</td>
<td>Inconsistent spellings between samples will be treated as separate
values. Application specific formatting such as colour coding and notes
in spreadsheets will be lost when converting to plain text formats.</td>
</tr>
<tr class="odd">
<td>5</td>
<td>
<strong>Numerical class values</strong> formatted as a factor (for
classification)</td>
<td>Some machine learning algorithms/libraries (e.g. SVM) require
classes defined by a number (e.g. -1 and +1 )</td>
</tr>
<tr class="even">
<td>6</td>
<td>
<strong>Balance of classes</strong> in the target variable (for
classification)</td>
<td>Imbalanced datasets may negatively impact results of machine
learning algorithms</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p><br></p>
<p>We’ll now apply these steps sequentially to the sample information
for the IBD dataset, contained in our variable
<code>samp.info.ibd</code>. First let’s take a look at the data:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 32
$ `Source Name`                             &lt;chr&gt; "Sample 1", "Sample 2", "Sam…
$ `Characteristics[organism]`               &lt;chr&gt; "Homo sapiens", "Homo sapien…
$ `Characteristics[age]`                    &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 2…
$ `Unit[time unit]`                         &lt;chr&gt; "year", "year", "year", "yea…
$ `Term Source REF`                         &lt;chr&gt; "EFO", "EFO", "EFO", "EFO", …
$ `Term Accession Number`                   &lt;chr&gt; "UO_0000036", "UO_0000036", …
$ `Characteristics[developmental stage]`    &lt;chr&gt; "adult", "adult", "adult", "…
$ `Characteristics[sex]`                    &lt;chr&gt; "male", "female", "male", "m…
$ `Characteristics[individual]`             &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 1…
$ `Characteristics[disease]`                &lt;chr&gt; "Crohns disease\tblood\torga…
$ `Characteristics[organism part]`          &lt;chr&gt; "", "blood", "blood", "blood…
$ `Material Type`                           &lt;chr&gt; "", "organism part", "organi…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117946", "P-MTAB…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117947", "P-MTAB…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117948", "P-MTAB…
$ `Extract Name`                            &lt;chr&gt; "", "Sample 2", "Sample 3", …
$ `Material Type`                           &lt;chr&gt; "", "RNA", "RNA", "RNA", "RN…
$ `Comment[LIBRARY_LAYOUT]`                 &lt;chr&gt; "", "PAIRED", "PAIRED", "PAI…
$ `Comment[LIBRARY_SELECTION]`              &lt;chr&gt; "", "other", "other", "other…
$ `Comment[LIBRARY_SOURCE]`                 &lt;chr&gt; "", "TRANSCRIPTOMIC", "TRANS…
$ `Comment[LIBRARY_STRATEGY]`               &lt;chr&gt; "", "Targeted-Capture", "Tar…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117950", "P-MTAB…
$ Performer                                 &lt;chr&gt; "", "Wellcome Trust Clinical…
$ `Assay Name`                              &lt;chr&gt; "", "Sample 2", "Sample 3", …
$ `Technology Type`                         &lt;chr&gt; "", "sequencing assay", "seq…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117949", "P-MTAB…
$ `Derived Array Data File`                 &lt;chr&gt; "", "ArrayExpress-normalized…
$ `Comment [Derived ArrayExpress FTP file]` &lt;chr&gt; "", "ftp://ftp.ebi.ac.uk/pub…
$ `Protocol REF`                            &lt;chr&gt; "", "P-MTAB-117949", "P-MTAB…
$ `Derived Array Data File`                 &lt;chr&gt; "", "ArrayExpress-raw.csv", …
$ `Comment [Derived ArrayExpress FTP file]` &lt;chr&gt; "", "ftp://ftp.ebi.ac.uk/pub…
$ `Factor Value[disease]`                   &lt;chr&gt; "", "normal", "normal", "nor…</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge 1:<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>On first glance, which of our checklist items do you think could be
an issue in the sample information for this dataset?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>All of the checklist items apply in this case. Let’s go through
them…</p>
</div>
</div>
</div>
</div>
<p><br></p>
<ol style="list-style-type: decimal">
<li>Let’s verify that there is a unique identifier that matches between
the sample information and counts matrix, and identify the name of the
column in the sample information. We manually find the names of the
samples in <code>raw.counts.ibd</code> (the counts matrix), and set to a
new variable <code>ibd.samp.names</code>. We then write a
<code>for loop</code> that evaluates which columns in the sample
information match the sample names. There is one match, the column named
<code>Source Name</code>. Note that there are other columns such as
<code>Assay Name</code> in this dataset that contain identifiers for
some but not all samples. This is a good illustration of why it
important to check carefully to ensure you have a complete set of unique
identifiers.</li>
</ol>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ibd.samp.names</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">raw.counts.ibd</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'raw.counts.ibd' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lst.colnames</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_along</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">lst.colnames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">all</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span> <span class="op">==</span> <span class="va">ibd.samp.names</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'ibd.samp.names' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"The unique IDs that match the counts matrix are in column: %s"</span>, <span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.ibd</span><span class="op">)</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">lst.colnames</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in which(lst.colnames): argument to 'which' is not logical</code></pre>
</div>
<p><br></p>
<ol start="2" style="list-style-type: decimal">
<li>We next extract this unique identifier along with the condition of
interest, (in this case the <code>disease</code>), and any other
variables that may be potential predictor variables or used to assess
confounding factors in the subsequent analysis of the data (in this case
patient <code>sex</code> and <code>age</code>). We will use functions
from the <code>dplyr</code> package, part of the <code>tidyverse</code>
collection of packages throughout this episode.</li>
</ol>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">require</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: tidyverse</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.2     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.2     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.1     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">samp.info.ibd</span>,</span>
<span>                              <span class="st">'Source Name'</span>,</span>
<span>                              <span class="st">'Characteristics[age]'</span>,</span>
<span>                              <span class="st">'Characteristics[sex]'</span>,</span>
<span>                              <span class="st">'Characteristics[disease]'</span></span>
<span>                              <span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="3" style="list-style-type: decimal">
<li>Let’s then rename the variables to something more easy to interpret
for humans, avoiding spaces and special characters. We’ll save these
selected columns to a new variable name
<code>samp.info.ibd.sel</code>.</li>
</ol>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>,</span>
<span>                          <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Source Name'</span>,</span>
<span>                          <span class="st">'age'</span> <span class="op">=</span> <span class="st">'Characteristics[age]'</span>,</span>
<span>                          <span class="st">'sex'</span> <span class="op">=</span> <span class="st">'Characteristics[sex]'</span>,</span>
<span>                          <span class="st">'condition'</span> <span class="op">=</span> <span class="st">'Characteristics[disease]'</span></span>
<span>                          <span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="4" style="list-style-type: decimal">
<li>Now check how each of the variables are encoded in our reduced
sample information data to identify any errors, data gaps and
inconsistent coding of categorical variables. Firstly let’s take a look
at the data.</li>
</ol>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 4
$ sampleID  &lt;chr&gt; "Sample 1", "Sample 2", "Sample 3", "Sample 4", "Sample 5", …
$ age       &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 24, 39, 41, 38, 46, 7, 46, 21, 35…
$ sex       &lt;chr&gt; "male", "female", "male", "male", "female", "female", "femal…
$ condition &lt;chr&gt; "Crohns disease\tblood\torganism part\tP-MTAB-117946\tP-MTAB…</code></pre>
</div>
<p><br></p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge 2:<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<p>What steps do we need to take to ensure that these four data columns
are consistently formatted and machine readable? Are you able to write
the code to address these issues?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ul>
<li>There is clearly an issue with the coding of the
<code>condition</code> Crohn’s disease. We’ll fix that first, and then
check the consistency of the coding of categorical variables
<code>sex</code> and <code>condition</code>.</li>
</ul>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">[</span><span class="fu">agrep</span><span class="op">(</span><span class="st">"Crohns"</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"crohns_disease"</span></span>
<span></span>
<span><span class="fu">unique</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">sex</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "male"               "female"             "crohns_disease"    
[4] "normal"             "ulcerative colitis"</code></pre>
</div>
<ul>
<li>Both <code>sex</code> and <code>condition</code> are consistently
encoded over all samples. We should however remove the spaces in the
values for <code>sampleID</code> and for the <code>condition</code>
value ulcerative colitis. Note: since we are changing our unique
identifier, we’ll need to make the same update later in the counts
matrix.</li>
</ul>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="va">samp.info.ibd.sel</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>sampleID <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">sampleID</span><span class="op">)</span>,</span>
<span>                                      condition <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="5" style="list-style-type: decimal">
<li>As some machine learning classification algorithms such as support
vector machines require a numerical input for each class, we’ll add in a
new numerical column to represent our target variable called
<code>class</code> where all patients denoted as <code>normal</code> are
given the value -1, and all patients denoted as either
<code>crohns_disease</code> or <code>ulcerative_colitis</code> are given
the vale +1. We’ll also now convert the type of categorical columns to
factors as this is preferred by downstream libraries used for data
normalisation such as DESeq2.</li>
</ol>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span> <span class="op">&lt;-</span> <span class="va">samp.info.ibd.sel</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>class <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">'normal'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'sex'</span>, <span class="st">'condition'</span>, <span class="st">'class'</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="st">'sex'</span>, <span class="st">'condition'</span>, <span class="st">'class'</span><span class="op">)</span><span class="op">]</span>, <span class="va">factor</span><span class="op">)</span></span></code></pre>
</div>
<p><br></p>
<ol start="6" style="list-style-type: decimal">
<li>Finally, let’s check the distribution of the classes by creating a
<code>table</code> from the <code>class</code> column.</li>
</ol>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">class</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
 -1   1 
267 323 </code></pre>
</div>
<p><br></p>
<p>The two classes are approximately equally represented, so let’s check
everything one last time.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">glimpse</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 590
Columns: 5
$ sampleID  &lt;chr&gt; "Sample_1", "Sample_2", "Sample_3", "Sample_4", "Sample_5", …
$ age       &lt;int&gt; 34, 49, 27, 9, 34, 60, 54, 24, 39, 41, 38, 46, 7, 46, 21, 35…
$ sex       &lt;fct&gt; male, female, male, male, female, female, female, male, male…
$ condition &lt;fct&gt; crohns_disease, normal, normal, normal, normal, ulcerative_c…
$ class     &lt;fct&gt; 1, -1, -1, -1, -1, 1, 1, -1, 1, 1, -1, 1, -1, -1, -1, 1, -1,…</code></pre>
</div>
<p><br></p>
<p>And save the output as a file.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">samp.info.ibd.sel</span>, file<span class="op">=</span><span class="st">"./data/ibd.sample.info.txt"</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="idb-dataset-counts-matrix-clean-up"><h2 class="section-heading">IDB dataset: Counts Matrix Clean Up<a class="anchor" aria-label="anchor" href="#idb-dataset-counts-matrix-clean-up"></a>
</h2>
<hr class="half-width">
<div id="count-matrix-readiness-checklist" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="count-matrix-readiness-checklist" class="callout-inner">
<h3 class="callout-title">Count Matrix Readiness Checklist<a class="anchor" aria-label="anchor" href="#count-matrix-readiness-checklist"></a>
</h3>
<div class="callout-content">
<table class="table">
<colgroup>
<col width="7%">
<col width="42%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Item</th>
<th>Check For…</th>
<th>Rationale</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>
<strong>Extraneous information</strong> that you don’t want to
include as a predictor variable</td>
<td>Blank rows, alternative gene/ transcript IDs and metadata about the
genes/transcripts will be treated as a predictor variable unless you
explicitly remove them</td>
</tr>
<tr class="even">
<td>2</td>
<td>
<strong>Unique variable names</strong>, without spaces or special
characters</td>
<td>ML algorithms require unique variables. Spaces etc. may cause errors
with bioinformatics and ml libraries used downstream (and they must
match the sample information)</td>
</tr>
<tr class="odd">
<td>3</td>
<td><strong>Duplicate data</strong></td>
<td>Duplicate variables may bias results of analysis and should be
removed unless intentional</td>
</tr>
<tr class="even">
<td>4</td>
<td><strong>Missing values</strong></td>
<td>Missing values may cause errors and bias results. Missing values
must be identified and appropriate treatment determined (e.g. drop vs
impute)</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p><br></p>
<ol style="list-style-type: decimal">
<li>Take a look at a sample of columns and rows to see what the
downloaded file looks like</li>
</ol>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw.counts.ibd</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'raw.counts.ibd' not found</code></pre>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge 3:<a class="anchor" aria-label="anchor" href="#challenge-3"></a>
</h3>
<div class="callout-content">
<p>Can you see the irrelevant information that we need to remove from
the counts matrix right away?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<ul>
<li>Remove the first row, which is in this case totals of counts for
each sample</li>
<li>Remove the first columns, which just numbers the transcript IDs</li>
<li>Move the column named <code>read</code> that contains to the
transcript IDs to the row names</li>
</ul>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.idb</span> <span class="op">&lt;-</span> <span class="va">raw.counts.ibd</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'raw.counts.ibd' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'counts.mat.idb' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.idb</span> <span class="op">&lt;-</span>  <span class="va">counts.mat.idb</span> <span class="op">%&gt;%</span> <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'read'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.idb</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="2" style="list-style-type: decimal">
<li>Let’s check for duplicate sampleIDs and transcript IDs. Provided
there are no duplicate row or column names, the following should return
<code>interger(0)</code>.</li>
</ol>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which</span><span class="op">(</span><span class="fu">duplicated</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which</span><span class="op">(</span><span class="fu">duplicated</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<p><br></p>
<ol start="3" style="list-style-type: decimal">
<li>We’ll double check the sampleIDs match the sample information
file.</li>
</ol>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu">identical</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span>, <span class="va">samp.info.ibd.sel</span><span class="op">$</span><span class="va">sampleID</span><span class="op">)</span><span class="op">)</span><span class="op">{</span><span class="kw">stop</span><span class="op">(</span><span class="op">)</span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4:<a class="anchor" aria-label="anchor" href="#challenge-4"></a>
</h3>
<div class="callout-content">
<p>Why did we get an error here?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>We renamed the samples in the sample information to remove spaces, so
we need to do the same here.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span>x <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"\ "</span>, replacement <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
<ol start="4" style="list-style-type: decimal">
<li>Finally, we’ll check that there are no missing values in the counts
matrix. Note, there will be many zeros, but we are ensuring that there
are no <code>NA</code> values or black records. The following should
return <code>FALSE</code>.</li>
</ol>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allMissValues</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu">all</span><span class="op">(</span><span class="fu">is.na</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">|</span> <span class="va">x</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span><span class="op">}</span></span>
<span></span>
<span><span class="fu">allMissValues</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<p><br></p>
<p>Take a final look at the cleaned up matrix.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.mat.idb</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i rows, corresponding to the transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"There are %i columns, corresponding to the samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">counts.mat.idb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'counts.mat.idb' not found</code></pre>
</div>
<p><br></p>
<p>And save the output as a file.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">write.table</span><span class="op">(</span><span class="va">counts.mat.idb</span>, file<span class="op">=</span><span class="st">"./data/counts.mat.idb.txt"</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div id="document-your-reformatting-steps" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="document-your-reformatting-steps" class="callout-inner">
<h3 class="callout-title">Document your reformatting steps<a class="anchor" aria-label="anchor" href="#document-your-reformatting-steps"></a>
</h3>
<div class="callout-content">
<p>It is crucial that you document your data reformatting so that it is
reproducible by other researchers. Create a project folder for your data
processing and include:</p>
<ul>
<li>Raw input data files collected from ArrayExpress, GEO or
similar</li>
<li>The output counts matrix and target variable vector, stored in a
common format such at .txt format</li>
<li>The full R scripts used to perform all the data clean up steps.
Store the script with the input and output data.</li>
<li>A readme file that explains the steps taken, and how to run the
script on the inputs to generate the outputs. This should also include
details of all software versions used (e.g. R version, RStudio version,
and the versions of any code libraries used)</li>
</ul>
</div>
</div>
</div>
</section><section id="the-tb-dataset-another-example"><h2 class="section-heading">The TB Dataset: Another Example<a class="anchor" aria-label="anchor" href="#the-tb-dataset-another-example"></a>
</h2>
<hr class="half-width">
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>Take a look at <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-6845" class="external-link">The
Tuburculosis (TB) Dataset - E-MTAB-6845</a> that was included in the
lesson data zip file.</p>
<p>In your groups, read the sdrf file into R. Each take a look at the
data and then as a group, make a list of the potential issues with this
data based on the checklist above? What steps would need to be taken to
get them ready to train a machine learning classifier? Record your
findings in the shared document.</p>
<p>As a help, the code to read the file in from your data directory
is:</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.tb</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"./data/E-MTAB-6845.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">"\t"</span>, header<span class="op">=</span><span class="cn">T</span>, fill<span class="op">=</span><span class="cn">T</span>, check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>Here are a number of issues that need addressing in the TB
dataset:</p>
<ul>
<li>One important reformatting point in this dataset is that there are
two lines in the sdrf file per sample, as there is a line for each of
the read in the paired end reads. The only difference is the file name
given for the fastq file. To solve this we need to select the distinct
rows with the variables of interest. Check the issue this line of
code:</li>
</ul>
<p><code>which(duplicated(samp.info.tb$Comment[sample id])))</code></p>
<ul>
<li>There are missing values for some variables that may be useful as
predictor variables, (e.g. <code>Characteristics[sex]</code> missing
values are represented by double spaces). It would be better to recode
them as <code>NA</code> as this is universally understood to represent a
missing value. This is easily seem by running this:</li>
</ul>
<p><code>unique(samp.info.tb$</code>Characteristics[sex]<code>)</code></p>
<ul>
<li><p>A number of variables need renaming and special characters and
spaces removed. For example, the variable name
<code>progressor status</code> and the value <code>TB progressor</code>
of this variable use a space that could replaced with an
underscore</p></li>
<li><p>And perhaps most importantly of all, the dataset is highly
unbalanced, with 9 TB progressors and 351 non-progressors when you
account for duplicate data. This dataset is very unlikely to perform
well as a training dataset for a machine learning classifier!</p></li>
</ul>
<p><code>table(samp.info.tb$</code>Factor Value[progressor status
(median follow-up 1.9 years)]<code>)</code></p>
</div>
</div>
</div>
</div>
<div id="challenge-extension" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-extension" class="callout-inner">
<h3 class="callout-title">Challenge Extension<a class="anchor" aria-label="anchor" href="#challenge-extension"></a>
</h3>
<div class="callout-content">
<p>Can you correctly reorder the following code snippets to create a
pipeline to reformat the data in line with the sample information
checklist? Use the
<code>Characteristics[progressor status (median follow-up 1.9 years)]</code>
as the target variable.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span></span>
<span>  <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Comment[sample id]'</span>,</span>
<span>  <span class="st">'prog_status'</span> <span class="op">=</span> <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span></span>
<span>            <span class="st">'Comment[sample id]'</span>,</span>
<span>            <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span>  <span class="op">%&gt;%</span></span>
<span></span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">prog_status</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"non_progressor"</span>,<span class="st">"tb_progressor"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             class <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">class</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>   </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">recode</span><span class="op">(</span><span class="va">prog_status</span>,                                </span>
<span>              <span class="st">"non-progressor"</span> <span class="op">=</span> <span class="st">"non_progressor"</span>,</span>
<span>              <span class="st">"TB progressor"</span> <span class="op">=</span> <span class="st">"tb_progressor"</span><span class="op">)</span>, </span>
<span>              class <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="va">prog_status</span> <span class="op">==</span> <span class="st">'non_progressor'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>                 </span>
<span>              <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span></span>
<span><span class="va">samp.info.tb</span> <span class="op">%&gt;%</span></span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="op">)</span>     </span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Example code to clean up the sample information
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.tb</span> <span class="op">%&gt;%</span> </span>
<span>  </span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span></span>
<span>            <span class="st">'Comment[sample id]'</span>,</span>
<span>            <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span>  <span class="op">%&gt;%</span>             <span class="co"># select columns</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">rename</span><span class="op">(</span></span>
<span>        <span class="st">'sampleID'</span> <span class="op">=</span> <span class="st">'Comment[sample id]'</span>,</span>
<span>        <span class="st">'prog_status'</span> <span class="op">=</span> <span class="st">'Factor Value[progressor status (median follow-up 1.9 years)]'</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># rename them</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>                                                                  <span class="co"># remove duplicate rows</span></span>
<span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">recode</span><span class="op">(</span><span class="va">prog_status</span>,                                 <span class="co"># recode target variable</span></span>
<span>                    <span class="st">"non-progressor"</span> <span class="op">=</span> <span class="st">"non_progressor"</span>,</span>
<span>                    <span class="st">"TB progressor"</span> <span class="op">=</span> <span class="st">"tb_progressor"</span><span class="op">)</span>, </span>
<span>                    class <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="va">prog_status</span> <span class="op">==</span> <span class="st">'non_progressor'</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>                 <span class="co"># create numerical class</span></span>
<span>                    <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span></span>
<span>       <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span>prog_status <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">prog_status</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"non_progressor"</span>,<span class="st">"tb_progressor"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                     class <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">class</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>                          <span class="co"># set variables to factors</span></span>
<span></span>
<span>       <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">glimpse</span><span class="op">(</span><span class="op">)</span>                                                                     <span class="co"># view output</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 360
Columns: 3
$ sampleID    &lt;chr&gt; "PR123_S19", "PR096_S13", "PR146_S14", "PR158_S12", "PR095…
$ prog_status &lt;fct&gt; non_progressor, non_progressor, non_progressor, non_progre…
$ class       &lt;fct&gt; -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1…</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Machine learning algorithms require specific data input formats, and
for data to be consistently formatted across variables and samples. A
classification task for example requires a matrix of the value of each
input variable for each sample, and the class label for each
sample.</li>
<li>Clinical sample information is often messy, with inconsistent
formatting as a result of how it is collected. This applies to data
downloaded from public repositories.</li>
<li>You must carefully check all data for formatting and data integrity
issues that may negatively impact your downstream ml analysis.</li>
<li>Document your data reformatting and store the code pipeline used
along with the raw and reformatted data to ensure your procedure is
reproducible</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-episode5"><p>Content from <a href="episode5.html">Data Readiness: Technical Noise</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode5.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do technical artefacts in RNA-Seq data impact machine learning
algorithms?</li>
<li>How can technical artefacts such as low count genes and outlier read
counts be effectively removed from RNA-Seq data prior to analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the main types of technical noise in RNA-Seq data, namely
low count genes and influential outlier read counts, and recall why
these exist in raw datasets</li>
<li>Demonstrate how to remove these elements from the data standard R
libraries to further improve machine learning readiness of RNA-Seq count
data</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="under-development" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="under-development" class="callout-inner">
<h3 class="callout-title">Under Development<a class="anchor" aria-label="anchor" href="#under-development"></a>
</h3>
<div class="callout-content">
<p>This episode is still actively being developed</p>
</div>
</div>
</div>
<section id="technical-artefacts-in-rna-seq-data"><h2 class="section-heading">Technical Artefacts in RNA-Seq Data<a class="anchor" aria-label="anchor" href="#technical-artefacts-in-rna-seq-data"></a>
</h2>
<hr class="half-width"></section><section id="low-count-filtering"><h2 class="section-heading">Low Count Filtering<a class="anchor" aria-label="anchor" href="#low-count-filtering"></a>
</h2>
<hr class="half-width"></section><section id="outlier-read-count-filtering"><h2 class="section-heading">Outlier Read Count Filtering<a class="anchor" aria-label="anchor" href="#outlier-read-count-filtering"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>RNA-Seq read counts contain two main sources of technical ‘noise’
and are unlikely to represent true biological signals from the samples:
Low count genes/transcripts and influential outlier read counts.</li>
<li>Filtering out low read count and influential outlier genes /
transcripts removes potentially biasing variables without negatively
impacting the performance of downstream machine learning analysis.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-episode6"><p>Content from <a href="episode6.html">Data Readiness: Distribution and Scale</a></p>
<hr>
<p> Last updated on 2023-06-23 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode6.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Do I need to transform or rescale RNA-Seq data before inputting into
a machine learning algorithm?</li>
<li>What are the most appropriate transformations and how do these
depend on the particular machine learning algorithm being employed?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the main characteristics of the distribution of RNA-Seq
count data and how this impacts the performance of machine learning
algorithms</li>
<li>Recall the main types of data transformation commonly applied to
RNA-Seq count data to account for the skewedness, zero-inflation and
heteroskedasticity</li>
<li>Demonstrate how to apply these transformations in R</li>
<li>Recall the main types of data rescaling commonly applied to data
prior to inputing into machine learning algorithms, and which rescaling
are most appropriate / necessary for a range of commonly used
algorithms</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="under-development" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="under-development" class="callout-inner">
<h3 class="callout-title">Under Development<a class="anchor" aria-label="anchor" href="#under-development"></a>
</h3>
<div class="callout-content">
<p>This episode is still actively being developed</p>
</div>
</div>
</div>
<section id="rna-seq-counts-data-a-skewed-distribution"><h2 class="section-heading">RNA-Seq Counts Data: A Skewed Distribution<a class="anchor" aria-label="anchor" href="#rna-seq-counts-data-a-skewed-distribution"></a>
</h2>
<hr class="half-width"></section><section id="variance-stabilising-and-rlog-transformation"><h2 class="section-heading">Variance Stabilising and rlog Transformation<a class="anchor" aria-label="anchor" href="#variance-stabilising-and-rlog-transformation"></a>
</h2>
<hr class="half-width"></section><section id="standardisation-and-min-max-scaling"><h2 class="section-heading">Standardisation and Min-Max Scaling<a class="anchor" aria-label="anchor" href="#standardisation-and-min-max-scaling"></a>
</h2>
<hr class="half-width">
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>RNA-Seq read count data is heavily skewed with a large percentage of
zero values. The distribution is heteroskedastic, meaning the variance
depends on the mean.</li>
<li>Standard transformations such as the variance stabilising
transformation and rlog transformation are designed to make the
distribution of RNA-Seq data more Gaussian, and therefore more
appropriately distributed for many machine learning models. These
transformations are improvements to a simple log2 transformation in
particular for low count values.</li>
<li>Many machine learning algorithms require predictor variables to be
on the same scale. Standardisation (z-score) and min-max scaling are two
common techniques to rescale variables to the same scale. Tree based
models are less sensitive to scaling. Standardisation is appropriate for
linear models such as logistic regression and svm. min-max scaling is
most appropirate for neural network based models.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:parkinsone@cardiff.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.4" class="external-link">sandpaper (0.12.4)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "machine learning, RNA-Seq, data readiness, pre-processing, data management",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/aio.html",
  "identifier": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/aio.html",
  "dateCreated": "2023-05-23",
  "dateModified": "2023-06-23",
  "datePublished": "2023-06-23"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

