<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-Seq: Data Readiness for Machine Learning Applications: Data Collection: GEO</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../episode3.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-Seq: Data Readiness for Machine Learning Applications
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-Seq: Data Readiness for Machine Learning Applications
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-Seq: Data Readiness for Machine Learning Applications
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 22%" class="percentage">
    22%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 22%" aria-valuenow="22" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../episode3.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="episode1.html">1. Introduction: Machine Learning Ready RNA-Seq Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="episode2.html">2. Data Collection: ArrayExpress</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Data Collection: GEO
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#geo-data-objects">GEO Data Objects</a></li>
<li><a href="#geo-file-formats">GEO File Formats</a></li>
<li><a href="#searching-for-a-dataset-on-geo">Searching for a Dataset on GEO</a></li>
<li><a href="#illustrative-dataset-covid-19-dataset">Illustrative Dataset: Covid-19 Dataset</a></li>
<li><a href="#downloading-and-reading-into-r">Downloading and Reading into R</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="episode4.html">4. Data Readiness: Data Format and Integrity</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="episode5.html">5. Data Readiness: Technical Noise</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="episode6.html">6. Data Readiness: Distribution and Scale</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/episode2.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/episode4.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/episode2.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Data Collection:</a>
        <a class="chapter-link float-end" href="../instructor/episode4.html" rel="next">Next: Data Readiness: Data... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Data Collection: GEO</h1>
        <p> Last updated on 2023-07-14 | 
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode3.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 20 minutes </p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What format are processed RNA-Seq dataset stored on GEO?</li>
<li>How do I search for a dataset that meets my requirements on
GEO?</li>
<li>How do I use the R package <code>GEOquery</code> to download
datasets from GEO into R?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain the expected data and file formats for processed RNA-Seq
datasets on GEO</li>
<li>Demonstrate ability to correctly identify the subset of files
required to create a dataset for a simple supervised machine learning
task (e.g., binary classification)</li>
<li>Demonstrate ability to download and save processed RNA-Seq data
files from GEO suitable for a simple supervised machine learning task
(e.g., binary classification)</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="geo-data-objects"><h2 class="section-heading">GEO Data Objects<a class="anchor" aria-label="anchor" href="#geo-data-objects"></a>
</h2>
<hr class="half-width"><p>NCBI ‘gene expression omnibus’ , GEO, is one of the largest
repositories for functional genomics data. GEO data are stored as
individual samples, series of samples and curated datasets. Accession
numbers of GEO broadly fall into four categories:</p>
<ul><li>GSM - GEO Sample. These refer to a single sample.</li>
<li>GSE - GEO Series. These refer to a list of GEO samples that together
form a single experiment.</li>
<li>GDS - GEO Dataset. These are curated datasets containing summarised
information from the underlying GEO samples.</li>
</ul><p>There are many more datasets available as series than there are
curated GEO Datasets. For this lesson, we’ll focus on finding a GEO
Series as an illustrative dataset.</p>
</section><section id="geo-file-formats"><h2 class="section-heading">GEO File Formats<a class="anchor" aria-label="anchor" href="#geo-file-formats"></a>
</h2>
<hr class="half-width"><p>Data for RNA-Seq studies have three components on GEO (similar to
microarray studies):</p>
<ol style="list-style-type: decimal"><li><strong>Metadata</strong></li>
</ol><p>The metadata is equivalent to sample data relationship information
provided on ArrayExpress and refers to the descriptive information about
the overall study and individual samples. Metadata also contains
protocol information and references to the raw and processed data files
names for each sample. Metadata is stored in the following standard
formats on GEO:</p>
<ul><li>SOFT (Simple Omnibus Format in Text) is a line-based plain text
format</li>
<li>MINiML is an alternative to SOFT and offers an XML rendering of
SOFT</li>
<li>Series Matrix is a summary text files that include a tab delimited
value-matrix table</li>
</ul><p>The raw and processed counts data is typically not included in any of
the above files for an RNA-Seq experiement (whereas they may be included
with a microarray experiment).</p>
<ol start="2" style="list-style-type: decimal"><li><strong>Processed data files</strong></li>
</ol><p>Processed data (raw counts, noramlised counts) are typically provided
in a supplementary file, often a .gz archive of text files. Raw and
normalised counts may be provided as separate supplementary files.</p>
<ol start="3" style="list-style-type: decimal"><li><strong>Raw data files</strong></li>
</ol><p>Raw data files provide the original reads and quality scores and are
stored in the NIH’s <a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">Sequence
Read Archive (SRA)</a>.</p>
<p>Further information on how RNA-Seq experiments are stored on GEO is
given in the <a href="https://www.ncbi.nlm.nih.gov/geo/info/seq.html" class="external-link">submission
guidelines</a> on the GEO website.</p>
</section><section id="searching-for-a-dataset-on-geo"><h2 class="section-heading">Searching for a Dataset on GEO<a class="anchor" aria-label="anchor" href="#searching-for-a-dataset-on-geo"></a>
</h2>
<hr class="half-width"><p>Let’s begin on the <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO
home page</a>. GEO provides online search and filter tools. We search by
adding “rna-seq” as a keyword in the search box, and then apply the
following filters to the subsequent results page, and rank the results
by the number of samples in descending order:</p>
<p><br></p>
<table class="table"><thead><tr class="header"><th>Filter</th>
<th>Selection</th>
</tr></thead><tbody><tr class="odd"><td>Keywords.</td>
<td>“rna-seq”</td>
</tr><tr class="even"><td>Entry type</td>
<td>Series or DataSets</td>
</tr><tr class="odd"><td>Organism</td>
<td>homo sapiens</td>
</tr><tr class="even"><td>Sample count</td>
<td>From 100 to 1,000</td>
</tr></tbody></table><p><br></p>
<p>The full search string is:</p>
<p><code>rna-seq[All Fields] AND "Homo sapiens"[Organism] AND (("gse"[Filter] OR "gds"[Filter]) AND ("100"[n_samples] : "1000"[n_samples]))</code></p>
<p><br></p>
<p>The filtered results page should looks like this:</p>
<figure><img src="../fig/geo_home_filtered.jpg" alt="Screen Shot of GEO DataSets Search Results Tab Filtered" class="figure mx-auto d-block"><figcaption>GEO Filtered Results</figcaption></figure></section><section id="illustrative-dataset-covid-19-dataset"><h2 class="section-heading">Illustrative Dataset: Covid-19 Dataset<a class="anchor" aria-label="anchor" href="#illustrative-dataset-covid-19-dataset"></a>
</h2>
<hr class="half-width"><p>The data set GSE212041 relates to the case-control study of
neutrophils in Covid-19+ patients named “Longitudinal characterization
of circulating neutrophils uncovers phenotypes associated with severity
in hospitalized COVID-19 patients” and comprises longitudinal human
samples from 306 hospitalized COVID-19+ patients, 78 symptomatic
controls, and 8 healthy controls, each at multiple time points, with a
total of 781 samples. We’ll call it the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE212041" class="external-link">Covid-19
dataset</a>. Processed data is provided as raw counts and TPM normalised
counts in the Supplementary files.</p>
<p>Let’s look at some of the basic information on this dataset:</p>
<p><br></p>
<table class="table"><thead><tr class="header"><th>Data Field</th>
<th>Values</th>
</tr></thead><tbody><tr class="odd"><td>Sample count</td>
<td>781</td>
</tr><tr class="even"><td>Experimental Design</td>
<td>case control design</td>
</tr><tr class="odd"><td>Experimental factors</td>
<td>COVID+, COVID- symptomatic, healthy</td>
</tr></tbody></table></section><section id="downloading-and-reading-into-r"><h2 class="section-heading">Downloading and Reading into R<a class="anchor" aria-label="anchor" href="#downloading-and-reading-into-r"></a>
</h2>
<hr class="half-width"><p><br></p>
<div class="section level3">
<h3 id="using-geoquery-to-download-a-geo-series-as-an-expression-set">Using GEOquery to download a GEO series as an expression set<a class="anchor" aria-label="anchor" href="#using-geoquery-to-download-a-geo-series-as-an-expression-set"></a></h3>
<p>The function <code>getGEO()</code> from the <code>GEOquery</code>
library provides a convenient way to download GEO SOFT format data into
R. Here we will download the GEO Series GSE212041. This blog provides
additional information on <a href="https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/" class="external-link">reading
GEO SOFT files into R</a></p>
<p>If there is more than one SOFT file for a GEO Series,
<code>getGEO()</code> will return a list of datasets. Let’s download
GSE212041.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gse212041</span> <span class="op">&lt;-</span> <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEO</span><span class="op">(</span><span class="st">"GSE212041"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Setting options('download.file.method.GEOquery'='auto')</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Setting options('GEOquery.inmemory.gpl'=FALSE)</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Found 2 file(s)</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GSE212041</span><span class="op">-</span><span class="va">GPL18573_series_matrix.txt.gz</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GSE212041</span><span class="op">-</span><span class="va">GPL24676_series_matrix.txt.gz</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"Number of files downloaded: %i"</span>, <span class="fu">length</span><span class="op">(</span><span class="va">gse212041</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Number of files downloaded: 2"</code></pre>
</div>
<div id="challenge-1-how-many-samples-are-in-each-file" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-how-many-samples-are-in-each-file" class="callout-inner">
<h3 class="callout-title">Challenge 1: How many samples are in each
file?<a class="anchor" aria-label="anchor" href="#challenge-1-how-many-samples-are-in-each-file"></a>
</h3>
<div class="callout-content">
<p>Write the code to check that the number of samples in each file gives
us the total expected number of 781 samples.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">writeLines</span><span class="op">(</span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"file %i: %i samples"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fu">c</span><span class="op">(</span><span class="fu">dim</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>file 1: 16 samples
file 2: 765 samples</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="extracting-the-metadata">Extracting the Metadata<a class="anchor" aria-label="anchor" href="#extracting-the-metadata"></a></h3>
<p>We’ll extract the metadata for the larger dataset (765 samples) and
examine the column names to verify that the file contains the expected
metadata about the experiment.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samp.info.cov19</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">samp.info.cov19</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "title"                   "geo_accession"          
 [3] "status"                  "submission_date"        
 [5] "last_update_date"        "type"                   
 [7] "channel_count"           "source_name_ch1"        
 [9] "organism_ch1"            "characteristics_ch1"    
[11] "characteristics_ch1.1"   "characteristics_ch1.2"  
[13] "characteristics_ch1.3"   "characteristics_ch1.4"  
[15] "molecule_ch1"            "extract_protocol_ch1"   
[17] "extract_protocol_ch1.1"  "taxid_ch1"              
[19] "description"             "description.1"          
[21] "data_processing"         "data_processing.1"      
[23] "data_processing.2"       "data_processing.3"      
[25] "data_processing.4"       "platform_id"            
[27] "contact_name"            "contact_email"          
[29] "contact_phone"           "contact_laboratory"     
[31] "contact_institute"       "contact_address"        
[33] "contact_city"            "contact_state"          
[35] "contact_zip/postal_code" "contact_country"        
[37] "data_row_count"          "instrument_model"       
[39] "library_selection"       "library_source"         
[41] "library_strategy"        "supplementary_file_1"   
[43] "acuity.max:ch1"          "cell type:ch1"          
[45] "covid-19 status:ch1"     "patient category:ch1"   
[47] "time point:ch1"         </code></pre>
</div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="getting-the-counts-data">Getting the Counts Data<a class="anchor" aria-label="anchor" href="#getting-the-counts-data"></a></h3>
<p>If we use the <code>exprs()</code> function to extract the counts
data from the expression slot of the downloaded dataset, we’ll see that
in this data series, the counts matrix is not there. The expression set
only contains a list of the accession numbers of the samples included in
the expression set, but not the actual count data. (The code below
attempts to view a sample of the data).</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu">exprs</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     GSM6507615 GSM6507616 GSM6507617 GSM6507618 GSM6507619 GSM6507620
     GSM6507621 GSM6507622 GSM6507623 GSM6507624</code></pre>
</div>
<p>We can verify this by looking at the dimensions of the object in the
exprs slot.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu">exprs</span><span class="op">(</span><span class="va">gse212041</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]   0 765</code></pre>
</div>
<p>To get the counts data, we need to download the Supplementary file
(not downloaded by the <code>getGeo()</code> function). The
<code>GEOquery::getGEOSuppFiles()</code> function enables us to view and
download supplementary files attached to a GEO Series (GSE). This
function doesn’t parse the downloaded files, since the file format may
vary between data series.</p>
<p>Set the ‘fetch_files’ argument to FALSE initially to view the
available files.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE212041"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                          <span class="op">)</span><span class="op">$</span><span class="va">fname</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
[1] "GSE212041_Neutrophil_RNAseq_Count_Matrix.txt.gz" "GSE212041_Neutrophil_RNAseq_TPM_Matrix.txt.gz" 
</code></pre>
</div>
<div id="challenge-2-download-the-supplementary-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2-download-the-supplementary-files" class="callout-inner">
<h3 class="callout-title">Challenge 2: Download the supplementary
files<a class="anchor" aria-label="anchor" href="#challenge-2-download-the-supplementary-files"></a>
</h3>
<div class="callout-content">
<p>Now Download the file to your /data subdirectory, by setting
<code>fetch_files</code> to TRUE and apply a <code>filter_regex</code>
of <code>'*Count_Matrix*'</code> to include the raw counts file only.
Read a sample of the file into R and check the dimensions of the file
correspond to the expected number of samples.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE212041"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="st">'*Count_Matrix*'</span></span>
<span>                          <span class="op">)</span> </span>
<span></span>
<span><span class="va">raw.counts.cov19</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/GSE212041_Neutrophil_RNAseq_Count_Matrix.txt.gz"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">"\t"</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i rows, corresponding to transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.cov19</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i columns, corresponding to samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.cov19</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
[1] "60640 rows, corresponding to transcript IDs"
[1] "783 columns, corresponding to samples"
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-3-download-dataset-gse157657-from-geo" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3-download-dataset-gse157657-from-geo" class="callout-inner">
<h3 class="callout-title">Challenge 3: Download dataset GSE157657 from
GEO<a class="anchor" aria-label="anchor" href="#challenge-3-download-dataset-gse157657-from-geo"></a>
</h3>
<div class="callout-content">
<p>Use the functions from the <code>GEOquery</code> library to download
the counts data and sample information for <code>GSE157657</code>, a
dataset of 760 TB patients.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gse157657</span> <span class="op">&lt;-</span> <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEO</span><span class="op">(</span><span class="st">"GSE157657"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samp.info.geotb</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">gse157657</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE157657"</span>,</span>
<span>                          makeDirectory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          baseDir <span class="op">=</span> <span class="st">"./data"</span>,</span>
<span>                          fetch_files <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          filter_regex <span class="op">=</span> <span class="cn">NULL</span></span>
<span>                          <span class="op">)</span> </span>
<span></span>
<span><span class="va">raw.counts.geotb</span> <span class="op">&lt;-</span> <span class="fu">read.table</span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/GSE157657_norm.data.txt.gz"</span>,</span>
<span>                             sep<span class="op">=</span><span class="st">"\t"</span>,</span>
<span>                             header<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             fill<span class="op">=</span><span class="cn">T</span>,</span>
<span>                             check.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i rows, corresponding to transcript IDs"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.geotb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">sprintf</span><span class="op">(</span><span class="st">"%i columns, corresponding to samples"</span>, <span class="fu">dim</span><span class="op">(</span><span class="va">raw.counts.geotb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "58735 rows, corresponding to transcript IDs"
[1] "761 columns, corresponding to samples"
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge 4:<a class="anchor" aria-label="anchor" href="#challenge-4"></a>
</h3>
<div class="callout-content">
<p>How have the counts data been processed in this dataset?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1">
  Give me a hint
  </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>Hint: you’ll need to refer back to the GEO website to understand how
the data has been processed.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>The counts data have been normalised and transformed using a variance
stabilising transformation using the <code>DESeq2</code> R library.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Similar to ArrayExpress, GEO stores samples information and counts
matrices separately. Sample information is typically stored in
<code>SOFT</code> format, as well as a <code>.txt</code> file.</li>
<li>Counts data may be stored as raw counts and some further processed
form of counts, typically as supplementary files. Always review the
documentation to determine what “processed data” refers to for a
particular dataset.</li>
<li>
<code>GEOquery</code> provides a convenient way to download files
directly from GEO into R.</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/episode2.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/episode4.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/episode2.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Data Collection:</a>
        <a class="chapter-link float-end" href="../instructor/episode4.html" rel="next">Next: Data Readiness: Data... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/edit/main/episodes/episode3.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/rna-seq-data-for-ml.git/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:parkinsone@cardiff.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.12.5)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/episode3.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "machine learning, RNA-Seq, data readiness, pre-processing, data management",
  "name": "Data Collection: GEO",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/episode3.html",
  "identifier": "https://carpentries-incubator.github.io/rna-seq-data-for-ml.git/instructor/episode3.html",
  "dateCreated": "2023-05-23",
  "dateModified": "2023-07-14",
  "datePublished": "2023-07-25"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

